<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="An R package to read in, manipulate, and visualize the outputs of sourmash.">
<title>Working with the Outputs of Sourmash in R • sourmashconsumr</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Working with the Outputs of Sourmash in R">
<meta property="og:description" content="An R package to read in, manipulate, and visualize the outputs of sourmash.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">sourmashconsumr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/sourmashconsumr.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="sourmashconsumr">sourmashconsumr<a class="anchor" aria-label="anchor" href="#sourmashconsumr"></a>
</h1></div>
<!-- badges: start -->

<p>The goal of sourmashconsumr is to parse, analyze, and visualize the outputs of the <a href="https://sourmash.readthedocs.io/en/latest/" class="external-link">sourmash python package</a>. The sourmashconsumr package is still under active development.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of sourmashconsumr from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("remotes")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"Arcadia-Science/sourmashconsumr"</span><span class="op">)</span></span></code></pre></div>
<p>Eventually, we hope to release sourmashconsumr on CRAN and to provide a conda-forge package. We’ll update these instructions once we’ve done that.</p>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>See the vignette for full instructions on how to run the sourmashconsumr package (coming soon!).</p>
<p>To access the functions in the sourmashconsumr package, you can load it with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://arcadia-science.github.io/sourmashconsumr/">sourmashconsumr</a></span><span class="op">)</span></span></code></pre></div>
<p>The sourmashconsumr package contains a variety of functions to work with the outputs of the sourmash python package. <code>read*</code> functions read and parse the outputs of sourmash into data frames. <code>plot*</code> functions visualize the sourmash outputs. <code>from_*to_*</code> functions convert sourmash data frames into the formats of other popular R packages (e.g. phyloseq). sourmashconsumr also contains a few utility functions that work with other datatypes (at this time, mostly upset data frames). See a complete list of functions below, sorted by sourmash output that the function operates on:</p>
<ul>
<li>signatures (output by <code>sourmash sketch</code> or <code>sourmash compute</code>):
<ul>
<li><code><a href="reference/read_signature.html">read_signature()</a></code></li>
<li>upset plots: <code><a href="reference/from_signatures_to_upset_df.html">from_signatures_to_upset_df()</a></code>, <code><a href="reference/plot_signatures_upset.html">plot_signatures_upset()</a></code>
</li>
<li>rarefaction plots for signatures sketched from reads: <code><a href="reference/from_signatures_to_rarefaction_df.html">from_signatures_to_rarefaction_df()</a></code>, <code><a href="reference/plot_signatures_rarefaction.html">plot_signatures_rarefaction()</a></code>
</li>
</ul>
</li>
<li>sourmash compare csv:
<ul>
<li><code>read_compare()</code></li>
<li>MDS plot: <code><a href="reference/make_compare_mds.html">make_compare_mds()</a></code>, <code><a href="reference/plot_compare_mds.html">plot_compare_mds()</a></code>
</li>
<li>heatmap: <code><a href="reference/plot_compare_heatmap.html">plot_compare_heatmap()</a></code>
</li>
</ul>
</li>
<li>sourmash taxonomy annotate csv:
<ul>
<li><code><a href="reference/read_taxonomy_annotate.html">read_taxonomy_annotate()</a></code></li>
<li>taxonomy agglomeration: <code><a href="reference/tax_glom_taxonomy_annotate.html">tax_glom_taxonomy_annotate()</a></code>
</li>
<li>upset plot: <code><a href="reference/from_taxonomy_annotate_to_upset_inputs.html">from_taxonomy_annotate_to_upset_inputs()</a></code>, <code><a href="reference/plot_taxonomy_annotate_upset.html">plot_taxonomy_annotate_upset()</a></code>
</li>
<li>sankey plot: <code><a href="reference/plot_taxonomy_annotate_sankey.html">plot_taxonomy_annotate_sankey()</a></code>
</li>
<li>time series alluvial plot: <code><a href="reference/plot_taxonomy_annotate_ts_alluvial.html">plot_taxonomy_annotate_ts_alluvial()</a></code>
</li>
<li>detect the presence of multiple strains of a single species in a metagenome: <code><a href="reference/from_taxonomy_annotate_to_multi_strains.html">from_taxonomy_annotate_to_multi_strains()</a></code>
</li>
<li>to metacoder: <code><a href="reference/from_taxonomy_annotate_to_metacoder.html">from_taxonomy_annotate_to_metacoder()</a></code>
</li>
<li>to phyloseq: <code><a href="reference/from_taxonomy_annotate_to_phyloseq.html">from_taxonomy_annotate_to_phyloseq()</a></code>
</li>
</ul>
</li>
<li>sourmash gather csv:
<ul>
<li><code><a href="reference/read_gather.html">read_gather()</a></code></li>
<li>barchart: <code><a href="reference/plot_gather_classified.html">plot_gather_classified()</a></code>
</li>
<li>upset plot: <code><a href="reference/from_gather_to_upset_df.html">from_gather_to_upset_df()</a></code>, <code><a href="reference/plot_gather_upset.html">plot_gather_upset()</a></code>
</li>
</ul>
</li>
<li>upset utilities:
<ul>
<li><code><a href="reference/from_list_to_upset_df.html">from_list_to_upset_df()</a></code></li>
<li><code><a href="reference/from_upset_df_to_intersection_members.html">from_upset_df_to_intersection_members()</a></code></li>
<li><code><a href="reference/from_upset_df_to_intersection_summary.html">from_upset_df_to_intersection_summary()</a></code></li>
<li><code><a href="reference/from_upset_df_to_intersections.html">from_upset_df_to_intersections()</a></code></li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="developer-documentation">Developer documentation<a class="anchor" aria-label="anchor" href="#developer-documentation"></a>
</h2>
<p>The sourmashconsumr package follows package developer conventions laid out in <a href="https://r-pkgs.org/" class="external-link uri">https://r-pkgs.org/</a>, and changes can be contributed to the code base using pull requests. For more information on how to contribute, see the <a href="devdoc.html">developer documentation</a>.</p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>sourmashconsumr doesn’t have a citation yet, but sourmash does! If you use sourmash in your work, please cite: <a href="https://joss.theoj.org/papers/10.21105/joss.00027" class="external-link">Brown and Irber (2016), doi:10.21105/joss.00027.</a></p>
<p>If you’d like more information on how sourmash works, please see the following publications:</p>
<ul>
<li>For a general background on how sourmash works and examples of how to use it: <a href="https://f1000research.com/articles/8-1006/v1" class="external-link">Large-scale sequence comparisons with sourmash</a>
</li>
<li>For a mathematical description of FracMinHash and a demonstration of the accuracy of sourmash gather: <a href="https://www.biorxiv.org/content/10.1101/2022.01.11.475838v2.abstract" class="external-link">Lightweight compositional analysis of metagenomes with FracMinHash and minimum metagenome covers</a>
</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing sourmashconsumr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Taylor Reiter <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Arcadia-Science/sourmashconsumr/actions/workflows/check-standard.yaml" class="external-link"><img src="https://github.com/Arcadia-Science/sourmashconsumr/actions/workflows/check-standard.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://app.codecov.io/gh/Arcadia-Science/sourmashconsumr?branch=main" class="external-link"><img src="https://codecov.io/gh/Arcadia-Science/sourmashconsumr/branch/main/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Taylor Reiter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
