---
title: "Using sourmashconsumr to parse, analyze, and visualize the outputs of the sourmash pythong package"
date: November 2022
output: 
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{Using sourmashconsumr to parse, analyze, and visualize the outputs of the sourmash pythong package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(sourmashconsumr)
```

```{r, eval = F, echo = F}
# note this block is not rendered in the readme as echo = F.
# build with:
# devtools::build_rmd("vignettes/sourmashconsumr.Rmd")
```

## Introduction

[Sourmash](https://sourmash.readthedocs.io/en/latest/) is a tool that facilitates lightweight and rapid comparisons between potentially very large sets of sequencing data (see [here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6720031/) and [here](https://www.biorxiv.org/content/10.1101/2022.01.11.475838v2.abstract)).

Sourmashconsumr provides parsing, analysis, and visualization functions for outputs generated by the sourmash commands `sketch`, `compare`, `gather`, and `taxonomy annotate`.
The sourmash outputs that sourmashconsumr works with are information-rich text files (either CSV or JSON). 
The sourmashconsumr functions are designed to help interpret these outputs.

sourmashconsumr does not re-implement any of the functionality in the sourmash python package. 

### How sourmash works

Sourmash uses FracMinHash sketching to generate compressed representations of sequencing data.
Two core parameters determine which sequences are represented in these sketches: k-mers size and scaled value.
The k-mer size dictates the length of sequences represented in the sketch; defaults are *k* = 21, *k* = 31, and *k* = 51.
While these sizes are somewhat empiric, overlap between related genomic sequences for a given k-mer size roughly approximates taxonomic relationships, with substantial overlap at *k* = 21 roughly approximating genus-level relatedness, *k* = 31 species-level, and *k* = 51 strain-level.
The scaled value dictates the fraction of k-mers in the original sequence that get included in the sketch.
Approximately 1/scaled value of k-mers in the original sequence get included in the sketch.
The sourmash defaults are 1000, 2000, and 10,000.
The scaled value takes advantage of the fact that a subset of data can be used to accurately estimate things like similarity and containment and dramatically sub-samples the original sequences thus facilitating rapid comparisons even between very large data sets.

### Sourmash outputs that sourmashconsumr works on

The sourmashconsumr package works with outputs from the following sourmash commands.
We provide a brief description of each command, but be sure to check the [sourmash documentation](https://sourmash.readthedocs.io/en/latest/) for a more complete explanation and examples.

+ `sourmash sketch`: Produces sketches for each sequence. Any sequencing data (DNA, protein, or 6-frame translation) in FASTQ or FASTA format can be sketched. Files produced by `sourmash sketch` are referred to as *signatures* because they may contain many sketches (one sketch for each k-mer size).
+ `sourmash compare`: Compares many sketches and produces a (dis)similarity matrix. If abundance information is ignored, the output represents jaccard distance. If abundance information is included, the output represents angular distance.
+ `sourmash gather`: Compares each query sketch against all sketches in a database to identify the minimum set of sketches that contain all of the known k-mers in the query. Typically, `sourmash gather` is run on metagenomes to determine all of the genomes that are contained within the sample, however `sourmash gather` can be used to assess containment between a query and a large database of sequences (e.g. for contamination discovery).
+ `sourmash taxonomy annotate`: Assigns taxonomic lineages to the sourmash gather matches.

## Installation

You can install the development version of sourmashconsumr from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("Arcadia-Science/sourmashconsumr")
```

Eventually, we hope to release sourmashconsumr on CRAN and to provide a conda-forge package. 
We'll update these instructions once we've done that.

To access the functions in the sourmashconsumr package:

``` r
library(sourmashconsumr)
```

## Quickstart 

## Problems & getting help

If you encounter any problems using sourmashconsumr, please post an issue on the issue tracker: https://github.com/Arcadia-Science/sourmashconsumr/issues.
If you encounter problems with sourmash, please post an issue on the sourmash issue tracker: https://github.com/sourmash-bio/sourmash/issues.

## Notes about the example data and package file paths

The sourmashconsumr package comes with a few example data sets that you can load to get a feel for data formats and to play around with the functions in the package.

You can access the data using the `data()` command.
Below we show this for the what the sourmash compare output looks like after it has been read into R using `read_compare_csv()`:

```{r example_data_compare}
data(gut_compare_df)
gut_compare_df
```

The other available data sets can be loaded with the following code:
```{r example_data}
data(gut_signatures_df)
data(gut_gather_df)
data(gut_taxonomy_annotate_df)
```

### Explanation about accessing external files packaged with sourmashconsumr

Because some of the functions read external files, we've also included example sourmash outputs with the package to demonstrate how the `read*()` functions work.
Specifying paths to data files packaged with an R package is a bit ugly...it takes advantage of the base R function `system.file` and information about the package to derive the file paths.
While this is helpful because it allows anyone to access the files in a package no matter where the package is installed thereby allowing anyone to run the code in the vignette, the syntax can be confusing because the example ends up looking a lot different than the code you would actually write. 
Whenever you see `system.file()` in this vignette, know that when you go to run similar code on your own sourmash outputs, you need to specify the path to your file to run the command without all of the accompanying `system.file()` code.
See an example below:

```r
# how we have to specify paths so that anyone can access the package data
fpath <- system.file("extdata", "SRR5947006.sig.gz", package="sourmashconsumr")
signature_df <- read_signature(fpath)
```

```r
# how you will specify a path when you actually run the code
signature_df <- read_signature("path/to/your_signature_file.sig")
```

### Origins of the example data

The example data are six samples from the [iHMP IBD](https://ibdmdb.org/) cohort. 
These are short shotgun metagenomes from stool microbiomes. 
While the iHMP was a longitudinal study, the samples in the example data are all time point 1 samples taken from different individuals. 
All individuals were symptomatic at time point 1, but three individuals were diagnosed with Crohnâ€™s disease (cd) at the end of the year, while three individuals were not (nonIBD).

| run_accession | group |
|---------------|-------|
| SRR5936131    |	cd    |
| SRR5947006    |	cd    |
| SRR5935765    |	cd    |
| SRR5936197    |	nonibd|
| SRR5946923    |	nonibd|
| SRR5946920    |	nonibd|

You can see the complete set of commands that were used to generate the data [here](ADD LINK ONCE PR IS MERGED).

## Complete start

### Working with sourmash signatures (FracMinHash sketches)

Sourmash signatures are compressed representations of sequencing data.
This data can be used for many different purposes such as [machine learning](https://www.biorxiv.org/content/10.1101/2022.06.30.498290v1), differential k-mer abundance analysis, and sequencing depth/rarefaction analysis.

#### Reading sourmash sigatures into R

To facilitate these types of analyses in R, the sourmashconsumr function `read_signature()` reads sourmash signatures produced by the `sourmash sketch` or `sourmash compute` function into a data frame.

```{r read_signature_one}
fpath <- system.file("extdata", "SRR5947006.sig.gz", package="sourmashconsumr")
signature_df <- read_signature(fpath)
head(signature_df)
```

If you have many signatures that you would like to read into the same data frame, you can use `purrr`:

```{r read_signature_many}
library(purrr)

# grab the files paths for sourmashconsumr example data
fpaths <- system.file("extdata", package = "sourmashconsumr") %>% list.files()
fpaths <- fpaths[grepl(pattern = ".sig.gz", x = fpaths)]

# read in the signature files
signature_df <- fpaths %>%
  map_dfr(read_signature)

head(signature_df)
```

#### Visualizing shared k-mer content between samples with an upset plot


* signatures (output by `sourmash sketch` or `sourmash compute`): 
    * `read_signature()`, show how to read multiple signatures using purrr,  
    * upset plots: `from_signatures_to_upset_df()`, `plot_signatures_upset()`
 

#### Assessing sample sequencing depth using rarefaction curves

It only makes sense to generate rarefaction curves from 
    * rarefaction plots for signatures sketched from reads: `from_signatures_to_rarefaction_df()`, `plot_signatures_rarefaction()`

### Working with the CSV output of sourmash compare

* sourmash compare csv: 
    * `read_compare()`
    *  MDS plot: `make_compare_mds()`, `plot_compare_mds()`
    * heatmap: `plot_compare_heatmap()`
    
### Working with the CSV output of sourmash gather

* sourmash gather csv
    * `read_gather()`
    * barchart: `plot_gather_classified()`
    * upset plot: `from_gather_to_upset_df()`, `plot_gather_upset()`
    
### Working with the CSV output of sourmash taxonomy annotate

* sourmash taxonomy annotate csv
    * `read_taxonomy_annotate()`
    * taxonomy agglomeration: `tax_glom_taxonomy_annotate()`
    * upset plot: `from_taxonomy_annotate_to_upset_inputs()`, `plot_taxonomy_annotate_upset()`
    * sankey plot:  `plot_taxonomy_annotate_sankey()`
    * time series alluvial plot: `plot_taxonomy_annotate_ts_alluvial()`
    * to metacoder: `from_taxonomy_annotate_to_metacoder()`
    * to phyloseq: `from_taxonomy_annotate_to_phyloseq()`
     
### Upset utility functions

The sourmashconsumr package includes utility functions to create and work with upset-formatted data frames.
These data frames record the presence-absence of the value that is intersected between many samples

* upset utilities
    * `from_list_to_upset_df()`
    * `from_upset_df_to_intersection_members()`
    * `from_upset_df_to_intersection_summary()`
    * `from_upset_df_to_intersections()`
    
SHOW THE HEAD OF AN UPSET DATAFRAME TO CONTEXTUALIZE THE FORMAT.


