<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="The results from running sourmash gather on each of six stool microbiome
metagenomes against the GTDB rs207 representatives database."><title>Gut microbiome sourmash gather data frame — gut_gather_df • sourmashconsumr</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Gut microbiome sourmash gather data frame — gut_gather_df"><meta property="og:description" content="The results from running sourmash gather on each of six stool microbiome
metagenomes against the GTDB rs207 representatives database."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sourmashconsumr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/sourmashconsumr.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Gut microbiome sourmash gather data frame</h1>
      
      <div class="d-none name"><code>gut_gather_df.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The results from running sourmash gather on each of six stool microbiome
metagenomes against the GTDB rs207 representatives database.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="va">gut_gather_df</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="format">Format<a class="anchor" aria-label="anchor" href="#format"></a></h2>
    <p>## `gut_gather_df`
A data frame with 1,062 rows and 30 columns:</p><dl><dt>intersect_bp</dt>
<dd><p>Numeric. Estimated number of intersected base pairs between a metagenome and a genome in a database.</p></dd>

  <dt>f_orig_query</dt>
<dd><p>Numeric. Fraction of the original query that belongs to the match.</p></dd>

  <dt>f_match</dt>
<dd><p>Numeric. Fraction of the matched genome in the leftover query.</p></dd>

  <dt>f_unique_to_query</dt>
<dd><p>Numeric. Fraction of the query that uniquely belongs to the match.</p></dd>

  <dt>f_unique_weighted</dt>
<dd><p>Numeric. Abundance-weighted fraction of the query that uniquely belongs to the match.</p></dd>

  <dt>average_abund</dt>
<dd><p>Numeric. Average abundance of k-mers in the metagenome that were in the match.</p></dd>

  <dt>median_abund</dt>
<dd><p>Numeric. Median abundance of k-mers in the metagenome that were in the match.</p></dd>

  <dt>std_abund</dt>
<dd><p>Numeric. Standard deviation of abundance of k-mers in the metagenome that were in the match.</p></dd>

  <dt>filename</dt>
<dd><p>Character. File path for the database on the computer that sourmash gather was executed on.</p></dd>

  <dt>name</dt>
<dd><p>Character. Name of matched genome in the sourmash gather database.</p></dd>

  <dt>genome_accession</dt>
<dd><p>Character. Genome accession solved by cutting of the name variable at the first space.</p></dd>

  <dt>md5</dt>
<dd><p>Character. MD5 hash for the matched genome sketch.</p></dd>

  <dt>f_match_orig</dt>
<dd><p>Numeric. Fraction of the matched genome in the original query prior to gather subtraction.</p></dd>

  <dt>unique_intersect_bp</dt>
<dd><p>Numeric. Estimated number of uniquely intersected base pairs between a metagenome and a genome in a database.</p></dd>

  <dt>gather_result_rank</dt>
<dd><p>Numeric. Rank of match in gather results.</p></dd>

  <dt>remaining_bp</dt>
<dd><p>Numeric. Remaining base pairs in the query after the match is removed.</p></dd>

  <dt>query_filename</dt>
<dd><p>Character. File name for the query derived from the query sketch.</p></dd>

  <dt>query_name</dt>
<dd><p>Character. Name of the query.</p></dd>

  <dt>query_md5</dt>
<dd><p>Character. MD5 hash for the query.</p></dd>

  <dt>query_bp</dt>
<dd><p>Character. Number of base pairs in the query.</p></dd>

  <dt>ksize</dt>
<dd><p>Character. K-mer size used for sourmash gather.</p></dd>

  <dt>moltype</dt>
<dd><p>Character. Molecule type used for sourmash gather.</p></dd>

  <dt>scaled</dt>
<dd><p>Numeric. Scaled value that sourmash gather was performed at.</p></dd>

  <dt>query_n_hashes</dt>
<dd><p>Numeric. Number of hashes (k-mers) in the query.</p></dd>

  <dt>query_abundance</dt>
<dd><p>Logical. Whether hash (k-mer) abundance information was a part of the query sketch.</p></dd>

  <dt>query_containment_ani</dt>
<dd><p>Numeric. Containment between the query and the match.</p></dd>

  <dt>match_containment_ani</dt>
<dd><p>Numeric. Containment between the match and the query.</p></dd>

  <dt>average_containment_ani</dt>
<dd><p>Numeric. Average of the two containment metrics query_containment_ani and match_containment_ani.</p></dd>

  <dt>max_containment_ani</dt>
<dd><p>Numeric. Maximum containment between the two containment metrics query_containment_ani and match_containment_ani.</p></dd>

  <dt>potential_false_negative</dt>
<dd><p>Logical. Whether the containment estimate is a potential false negative.</p></dd>

  ...

</dl></div>
    <div class="section level2">
    <h2 id="source">Source<a class="anchor" aria-label="anchor" href="#source"></a></h2>
    <p>&lt;https://github.com/Arcadia-Science/sourmashconsumr/blob/main/data-raw/00_sourmash_commands.sh&gt;</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Taylor Reiter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

