<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ sourmashconsumr</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sourmashconsumr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/sourmashconsumr.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="check_and_edit_names_in_signatures_df.html">check_and_edit_names_in_signatures_df()</a></code> 
        </dt>
        <dd>Check for missing name, NAs in name, or empty name in a data frame of signatures</dd>
      </dl><dl><dt>
          
          <code><a href="check_compare_df_sample_col_and_move_to_rowname.html">check_compare_df_sample_col_and_move_to_rowname()</a></code> 
        </dt>
        <dd>Internal function to check if the sample name is a column or a rowname and move it to a rowname if it is a column.</dd>
      </dl><dl><dt>
          
          <code><a href="check_file_and_url_one.html">check_file_and_url_one()</a></code> 
        </dt>
        <dd>Check if local file path or URL are valid</dd>
      </dl><dl><dt>
          
          <code><a href="check_uniform_parameters_in_signatures_df.html">check_uniform_parameters_in_signatures_df()</a></code> 
        </dt>
        <dd>Check uniformity in sketch parameters</dd>
      </dl><dl><dt>
          
          <code><a href="from_gather_to_upset_df.html">from_gather_to_upset_df()</a></code> 
        </dt>
        <dd>Convert a data frame containing sourmash gather results into a upset plot compatible data frame</dd>
      </dl><dl><dt>
          
          <code><a href="from_list_to_upset_df.html">from_list_to_upset_df()</a></code> 
        </dt>
        <dd>Convert a list of vectors into an upset plot-compatible data.frame while conserving value names as row.names</dd>
      </dl><dl><dt>
          
          <code><a href="from_signatures_to_rarefaction_df.html">from_signatures_to_rarefaction_df()</a></code> 
        </dt>
        <dd>Create a tidy data.frame representing per-sample rarefaction information</dd>
      </dl><dl><dt>
          
          <code><a href="from_signatures_to_upset_df.html">from_signatures_to_upset_df()</a></code> 
        </dt>
        <dd>Convert a data.frame containing hashes from many sourmash signatures into a upset plot compatible data.frame</dd>
      </dl><dl><dt>
          
          <code><a href="from_taxonomy_annotate_to_count_table.html">from_taxonomy_annotate_to_count_table()</a></code> 
        </dt>
        <dd>Transform the output of sourmash taxonomy annotate into a phyloseq OTU table (count table)</dd>
      </dl><dl><dt>
          
          <code><a href="from_taxonomy_annotate_to_metacoder.html">from_taxonomy_annotate_to_metacoder()</a></code> 
        </dt>
        <dd>Transform the output of sourmash taxonomy annotate into a taxmap metacoder object</dd>
      </dl><dl><dt>
          
          <code><a href="from_taxonomy_annotate_to_multi_strains.html">from_taxonomy_annotate_to_multi_strains()</a></code> 
        </dt>
        <dd>Detect whether multiple strains of the same species are present in a sample.</dd>
      </dl><dl><dt>
          
          <code><a href="from_taxonomy_annotate_to_phyloseq.html">from_taxonomy_annotate_to_phyloseq()</a></code> 
        </dt>
        <dd>Transform the output of sourmash taxonomy annotate into a phyloseq object</dd>
      </dl><dl><dt>
          
          <code><a href="from_taxonomy_annotate_to_tax_table.html">from_taxonomy_annotate_to_tax_table()</a></code> 
        </dt>
        <dd>Transform the output of sourmash taxonomy annotate into a phyloseq tax table</dd>
      </dl><dl><dt>
          
          <code><a href="from_taxonomy_annotate_to_upset_inputs.html">from_taxonomy_annotate_to_upset_inputs()</a></code> 
        </dt>
        <dd>Transform a taxonomy annotate data frame into an upset plot compliant data frame</dd>
      </dl><dl><dt>
          
          <code><a href="from_upset_df_to_intersection_members.html">from_upset_df_to_intersection_members()</a></code> 
        </dt>
        <dd>Extract the lineages that underlie the intersections in an upset plot</dd>
      </dl><dl><dt>
          
          <code><a href="from_upset_df_to_intersection_summary.html">from_upset_df_to_intersection_summary()</a></code> 
        </dt>
        <dd>Summarize the size of intersections between samples in an upset compliant data frame</dd>
      </dl><dl><dt>
          
          <code><a href="from_upset_df_to_intersections.html">from_upset_df_to_intersections()</a></code> 
        </dt>
        <dd>Transform an upset compliant data frame into a data frame of intersection names and lineages</dd>
      </dl><dl><dt>
          
          <code><a href="get_scaled_for_max_hash.html">get_scaled_for_max_hash()</a></code> 
        </dt>
        <dd>Calculate the scaled value for a signature from the max_hash value</dd>
      </dl><dl><dt>
          
          <code><a href="gut_compare_df.html">gut_compare_df</a></code> 
        </dt>
        <dd>Gut microbiome sourmash compare data frame</dd>
      </dl><dl><dt>
          
          <code><a href="gut_gather_df.html">gut_gather_df</a></code> 
        </dt>
        <dd>Gut microbiome sourmash gather data frame</dd>
      </dl><dl><dt>
          
          <code><a href="gut_signatures_df.html">gut_signatures_df</a></code> 
        </dt>
        <dd>Gut microbiome sourmash signatures data frame</dd>
      </dl><dl><dt>
          
          <code><a href="gut_taxonomy_annotate_df.html">gut_taxonomy_annotate_df</a></code> 
        </dt>
        <dd>Gut microbiome sourmash taxonomy annotate data frame</dd>
      </dl><dl><dt>
          
          <code><a href="make_agglom_cols.html">make_agglom_cols()</a></code> 
        </dt>
        <dd>Helper function to define column names for taxonomy agglomeration</dd>
      </dl><dl><dt>
          
          <code><a href="make_compare_mds.html">make_compare_mds()</a></code> 
        </dt>
        <dd>Perform an MDS analysis from a sourmash compare output and produce a tidy data.frame</dd>
      </dl><dl><dt>
          
          <code><a href="make_expression.html">make_expression()</a></code> 
        </dt>
        <dd>Make an expression vector for italicizing a subset of a vector</dd>
      </dl><dl><dt>
          
          <code><a href="pivot_wider_taxonomy_annotate.html">pivot_wider_taxonomy_annotate()</a></code> 
        </dt>
        <dd>Pivot sourmash_taxonomy_results to wide format</dd>
      </dl><dl><dt>
          
          <code><a href="plot_compare_heatmap.html">plot_compare_heatmap()</a></code> 
        </dt>
        <dd>Plot a heatmap from a tibble of sourmash compare results.</dd>
      </dl><dl><dt>
          
          <code><a href="plot_compare_mds.html">plot_compare_mds()</a></code> 
        </dt>
        <dd>Plot an MDS data.frame produced from the output of sourmash compare</dd>
      </dl><dl><dt>
          
          <code><a href="plot_gather_classified.html">plot_gather_classified()</a></code> 
        </dt>
        <dd>Visualize the classified and unclassified fraction of many samples using sourmash gather results.</dd>
      </dl><dl><dt>
          
          <code><a href="plot_gather_upset.html">plot_gather_upset()</a></code> 
        </dt>
        <dd>Visualize the intersection of genome accessions in sourmash gather results from many samples</dd>
      </dl><dl><dt>
          
          <code><a href="plot_signatures_rarefaction.html">plot_signatures_rarefaction()</a></code> 
        </dt>
        <dd>Plot a rarefaction curve to determine sequence saturation</dd>
      </dl><dl><dt>
          
          <code><a href="plot_signatures_upset.html">plot_signatures_upset()</a></code> 
        </dt>
        <dd>Visualize the intersection of minhashes in sourmash signatures from many samples</dd>
      </dl><dl><dt>
          
          <code><a href="plot_taxonomy_annotate_sankey.html">plot_taxonomy_annotate_sankey()</a></code> 
        </dt>
        <dd>Visualize a sankey diagram from taxonomic lineages from one or many samples</dd>
      </dl><dl><dt>
          
          <code><a href="plot_taxonomy_annotate_ts_alluvial.html">plot_taxonomy_annotate_ts_alluvial()</a></code> 
        </dt>
        <dd>Visualize an allivual flow plot from taxonomic lineages from one or many samples</dd>
      </dl><dl><dt>
          
          <code><a href="plot_taxonomy_annotate_upset.html">plot_taxonomy_annotate_upset()</a></code> 
        </dt>
        <dd>Visualize an upset plot of taxonomic lineage intersections between samples</dd>
      </dl><dl><dt>
          
          <code><a href="read_compare_csv.html">read_compare_csv()</a></code> 
        </dt>
        <dd>Read a CSV file output by sourmash compare</dd>
      </dl><dl><dt>
          
          <code><a href="read_gather.html">read_gather()</a></code> 
        </dt>
        <dd>Read CSV file or files output by sourmash gather</dd>
      </dl><dl><dt>
          
          <code><a href="read_signature.html">read_signature()</a></code> 
        </dt>
        <dd>Read sourmash signature or signatures into a dataframe</dd>
      </dl><dl><dt>
          
          <code><a href="read_signature_one.html">read_signature_one()</a></code> 
        </dt>
        <dd>Read one sourmash signature into a dataframe</dd>
      </dl><dl><dt>
          
          <code><a href="read_taxonomy_annotate.html">read_taxonomy_annotate()</a></code> 
        </dt>
        <dd>Read CSV file or files output by sourmash taxonomy annotate</dd>
      </dl><dl><dt>
          
          <code><a href="tax_glom_taxonomy_annotate.html">tax_glom_taxonomy_annotate()</a></code> 
        </dt>
        <dd>Agglomerate counts of same lineage to specified level of taxonomy.</dd>
      </dl><dl><dt>
          
          <code><a href="warning_handler.html">warning_handler()</a></code> 
        </dt>
        <dd>Warning handler to suppress warning message that contains "The following named parsers don't match the column names:"</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Taylor Reiter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

