<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="sourmashconsumr">
<title>Using sourmashconsumr to parse, analyze, and visualize the outputs of the sourmash python package • sourmashconsumr</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Using sourmashconsumr to parse, analyze, and visualize the outputs of the sourmash python package">
<meta property="og:description" content="sourmashconsumr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sourmashconsumr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/sourmashconsumr.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Using sourmashconsumr to parse, analyze, and visualize the outputs of the sourmash python package</h1>
            
            <h4 data-toc-skip class="date">November 2022</h4>
      
      
      <div class="d-none name"><code>sourmashconsumr.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><a href="https://sourmash.readthedocs.io/en/latest/" class="external-link">Sourmash</a> is
a tool that facilitates lightweight and rapid comparisons between
potentially very large sets of sequencing data<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6720031/" class="external-link uri"&gt;https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6720031/&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a> <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="https://www.biorxiv.org/content/10.1101/2022.01.11.475838v2.abstract" class="external-link uri"&gt;https://www.biorxiv.org/content/10.1101/2022.01.11.475838v2.abstract&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a>.</p>
<p>Sourmashconsumr provides parsing, analysis, and visualization
functions for outputs generated by the sourmash commands
<code>sketch</code>, <code>compare</code>, <code>gather</code>, and
<code>taxonomy annotate</code>. The sourmash outputs that
sourmashconsumr works with are information-rich text files (either CSV
or JSON). The sourmashconsumr functions are designed to help interpret
these outputs.</p>
<p>sourmashconsumr does not re-implement any of the functionality in the
original sourmash software.</p>
<div class="section level3">
<h3 id="how-sourmash-works">How sourmash works<a class="anchor" aria-label="anchor" href="#how-sourmash-works"></a>
</h3>
<p>Sourmash uses FracMinHash sketching to generate compressed
representations of sequencing data. Two core parameters determine which
sequences are represented in these sketches: k-mer size and scaled
value. The k-mer size dictates the length of sequences represented in
the sketch; defaults are <em>k</em> = 21, <em>k</em> = 31, and
<em>k</em> = 51. While these sizes are somewhat empiric, overlap between
related genomic sequences for a given k-mer size roughly approximates
taxonomic relationships, with substantial overlap at <em>k</em> = 21
roughly approximating genus-level relatedness, <em>k</em> = 31
species-level, and <em>k</em> = 51 strain-level <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="https://doi.org/10.1128/mSystems.00020-16" class="external-link uri"&gt;https://doi.org/10.1128/mSystems.00020-16&lt;/a&gt;&lt;/p&gt;'><sup>3</sup></a>. The scaled value
dictates the fraction of k-mers in the original sequence that get
included in the sketch. Approximately <code>1/scaled</code> value of
k-mers in the original sequence are retained in the sketch. The most
common scaled values are 1000, 2000, and 10,000. The scaled value takes
advantage of the fact that a subset of sequencing data can be used to
accurately estimate things like similarity and containment and
dramatically sub-samples the original sequences thus facilitating rapid
comparisons even between very large data sets.</p>
</div>
<div class="section level3">
<h3 id="sourmash-outputs-that-sourmashconsumr-works-on">Sourmash outputs that sourmashconsumr works on<a class="anchor" aria-label="anchor" href="#sourmash-outputs-that-sourmashconsumr-works-on"></a>
</h3>
<p>The sourmashconsumr package works with outputs from the following
sourmash commands. We provide a brief description of each command, but
be sure to check the <a href="https://sourmash.readthedocs.io/en/latest/" class="external-link">sourmash
documentation</a> for a more complete explanation and examples.</p>
<ul>
<li>
<code>sourmash sketch</code>: Produces sketches for each sequence.
Any sequencing data (DNA, protein, or 6-frame translation) in FASTQ or
FASTA format can be sketched. Files produced by
<code>sourmash sketch</code> are referred to as <em>signatures</em>
because they may contain many sketches (one sketch for each k-mer
size).</li>
<li>
<code>sourmash compare</code>: Compares many sketches and produces a
(dis)similarity matrix. If abundance information is ignored, the output
represents jaccard distance. If abundance information is included, the
output represents angular distance.</li>
<li>
<code>sourmash gather</code>: Compares each query sketch against all
sketches in a database to identify the minimum set of sketches that
contain all of the known k-mers in the query. Typically,
<code>sourmash gather</code> is run on metagenomes to determine all of
the genomes that are contained within the sample, however
<code>sourmash gather</code> can be used to assess containment between a
query and a large database of sequences (e.g. for contamination
discovery).</li>
<li>
<code>sourmash taxonomy annotate</code>: Assigns taxonomic lineages
to the sourmash gather matches.</li>
</ul>
<div class="warning" style="padding:0.1em; background-color:#F0F8FF; color:#000000">
<span>
<p style="margin-top:1em; text-align:left">
<b>Naming signatures in sourmash to get the most out of sourmashconsumr
functions</b>
</p>
<p style="margin-left:1em;">
</p>
<p>Many functions in the sourmashconsumr R package rely on signatures
names. Signature names are set when you sketch a file by using the
<code>--name</code> or <code>--merge</code> flags:</p>
<pre><code>sourmash sketch dna -p k=21,k=31,k=51,scaled=1000,abund -o sample1.sig --name sample1 sample1*fastq.gz</code></pre>
<p>The above command would create a signature with three DNA sketches
(one for each k size), each with a scaled value of 1000 and abundance
tracking turned on. The signature name would be <code>sample1</code> and
it would contain sequences from all FASTQ files that start with
<code>sample1</code> and end in <code>fastq.gz</code>, and it would be
written to the file <code>sample1.sig</code>.</p>
<p>The signature name would be read into R as a column,
<code>name</code>, if you read the output file <code>sample1.sig</code>
into R (see below). The signature name is also propagated to the outputs
of <code>sourmash gather</code> and
<code>sourmash taxonomy annotate</code>; if you read these output files
into R, the signature name would be in the column
<code>query_name</code>.</p>
<p>Many of the sourmashconsumr functions use the <code>name</code> or
<code>query_name</code> columns. If you didn’t use the
<code>--name</code> flag when you sketched your signatures, you have two
options to create the <code>name</code>/<code>query_name</code>
column:</p>
<ol style="list-style-type: decimal">
<li>If you would like to add names to your sourmash signatures, you can
use the sourmash command <code>sourmash sig rename</code>. This might be
the best solution if you plan to work with the signatures more.</li>
</ol>
<pre><code>sourmash signature rename file1.sig "new name" -o renamed.sig</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>If you already ran <code>sourmash gather</code> on your signatures,
you can instead edit this column in R or excel. For example, you could
do something like this in R:</li>
</ol>
<pre><code><span><span class="fu">sourmashconsumr</span><span class="fu">::</span><span class="fu"><a href="../reference/read_gather.html">read_gather</a></span><span class="op">(</span><span class="st">"my_gather_results.csv"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>query_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">query_filename</span><span class="op">)</span><span class="op">)</span></span></code></pre>

</span>
</div>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of sourmashconsumr from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("remotes")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"Arcadia-Science/sourmashconsumr"</span><span class="op">)</span></span></code></pre></div>
<p>Eventually, we hope to release sourmashconsumr on CRAN and to provide
a conda-forge package. We’ll update these instructions once we’ve done
that.</p>
<p>To access the functions in the sourmashconsumr package:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://arcadia-science.github.io/sourmashconsumr/">sourmashconsumr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="problems-getting-help">Problems &amp; getting help<a class="anchor" aria-label="anchor" href="#problems-getting-help"></a>
</h2>
<p>If you encounter any problems using sourmashconsumr, please post an
issue on the issue tracker: <a href="https://github.com/Arcadia-Science/sourmashconsumr/issues" class="external-link uri">https://github.com/Arcadia-Science/sourmashconsumr/issues</a>.
If you encounter problems with sourmash, please post an issue on the
sourmash issue tracker: <a href="https://github.com/sourmash-bio/sourmash/issues" class="external-link uri">https://github.com/sourmash-bio/sourmash/issues</a>.</p>
</div>
<div class="section level2">
<h2 id="package-functionality">Package functionality<a class="anchor" aria-label="anchor" href="#package-functionality"></a>
</h2>
<div class="section level3">
<h3 id="load-example-data-and-vignette-dependencies">Load example data and vignette dependencies<a class="anchor" aria-label="anchor" href="#load-example-data-and-vignette-dependencies"></a>
</h3>
<p>Load the packages this vignette uses:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://arcadia-science.github.io/sourmashconsumr/">sourmashconsumr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span></code></pre></div>
<p>Load the example data sets:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gut_compare_df</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gut_signatures_df</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gut_gather_df</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gut_taxonomy_annotate_df</span><span class="op">)</span></span></code></pre></div>
<p>Make a metadata frame that corresponds to the example data:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a metadata data frame </span></span>
<span><span class="va">run_accessions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SRR5936131"</span>, <span class="st">"SRR5947006"</span>, <span class="st">"SRR5935765"</span>,</span>
<span>                    <span class="st">"SRR5936197"</span>, <span class="st">"SRR5946923"</span>, <span class="st">"SRR5946920"</span><span class="op">)</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cd"</span>, <span class="st">"cd"</span>, <span class="st">"cd"</span>, <span class="st">"nonibd"</span>, <span class="st">"nonibd"</span>, <span class="st">"nonibd"</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>run_accessions <span class="op">=</span> <span class="va">run_accessions</span>, groups <span class="op">=</span> <span class="va">groups</span><span class="op">)</span> </span></code></pre></div>
<div class="section level4">
<h4 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h4>
<p>The sourmashconsumr package comes with a few example data sets that
you can load to get a feel for data formats and to play around with the
functions in the package.</p>
<p>The example data are six samples from the <a href="https://ibdmdb.org/" class="external-link">iHMP IBD</a> cohort. These are short shotgun
metagenomes from stool microbiomes. While the iHMP was a longitudinal
study, the samples in the example data are all time point 1 samples
taken from different individuals. All individuals were symptomatic at
time point 1, but three individuals were diagnosed with Crohn’s disease
(cd) at the end of the year, while three individuals were not
(nonIBD).</p>
<table class="table">
<thead><tr class="header">
<th>run_accession</th>
<th>group</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>SRR5936131</td>
<td>cd</td>
</tr>
<tr class="even">
<td>SRR5947006</td>
<td>cd</td>
</tr>
<tr class="odd">
<td>SRR5935765</td>
<td>cd</td>
</tr>
<tr class="even">
<td>SRR5936197</td>
<td>nonibd</td>
</tr>
<tr class="odd">
<td>SRR5946923</td>
<td>nonibd</td>
</tr>
<tr class="even">
<td>SRR5946920</td>
<td>nonibd</td>
</tr>
</tbody>
</table>
<p>You can see the complete set of commands that were used to generate
the data <a href="ADD%20LINK%20ONCE%20PR%20IS%20MERGED">here</a>.</p>
<p>We’ve documented the content of each of the test data sets, including
an explanation for what is encoded by each column. If you’re having
trouble understanding the content of the sourmash output files, the data
set documentation may help. Use the following code to access this
information:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">gut_gather_df</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="reading-sourmash-outputs-into-r">Reading sourmash outputs into R<a class="anchor" aria-label="anchor" href="#reading-sourmash-outputs-into-r"></a>
</h3>
<p>All of the sourmashconsumr functions that read in the sourmash
outputs start with <code>read_</code>. All functions accepts compressed
or uncompressed files and can read one or multiple files at a time. When
multiple file paths are passed to a function, all files are read into a
single data frame. If you would like to read in multiple files at once,
the <code><a href="https://rdrr.io/r/base/Sys.glob.html" class="external-link">Sys.glob()</a></code> function or the <code><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files()</a></code>
function are useful to determine the paths for many files.</p>
<div class="warning" style="padding:0.1em; background-color:#F0F8FF; color:#000000">
<span>
<p style="margin-top:1em; text-align:left">
<b>Explanation about accessing external files packaged with
sourmashconsumr</b>
</p>
<p style="margin-left:1em;">
</p>
<p>Because some of the functions read external files, we’ve also
included example sourmash outputs with the package to demonstrate how
the <code>read*()</code> functions work. Specifying paths to data files
packaged with an R package is a bit ugly…it takes advantage of the base
R function <code>system.file</code> and information about the package to
derive the file paths. While this is helpful because it allows anyone to
access the files in a package no matter where the package is installed
on their computer, the syntax can be confusing because the example ends
up looking a lot different than the code you would actually write.
Whenever you see <code><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file()</a></code> in this vignette, know that
when you go to run similar code on your own sourmash outputs, you need
to specify the path to your file to run the command without all of the
accompanying <code><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file()</a></code> code. See an example below:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># how we have to specify paths so that anyone can access the package data</span></span>
<span><span class="va">fpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"SRR5947006.sig.gz"</span>, package<span class="op">=</span><span class="st">"sourmashconsumr"</span><span class="op">)</span></span>
<span><span class="va">signature_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_signature.html">read_signature</a></span><span class="op">(</span><span class="va">fpath</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># how you will specify a path when you actually run the code</span></span>
<span><span class="va">signature_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_signature.html">read_signature</a></span><span class="op">(</span><span class="st">"path/to/your_signature_file.sig"</span><span class="op">)</span></span></code></pre></div>

</span>
</div>
<div class="section level4">
<h4 id="reading-sourmash-sigatures-into-r">Reading sourmash sigatures into R<a class="anchor" aria-label="anchor" href="#reading-sourmash-sigatures-into-r"></a>
</h4>
<p>The sourmashconsumr function <code><a href="../reference/read_signature.html">read_signature()</a></code> reads
sourmash signatures produced by the <code>sourmash sketch</code> or
<code>sourmash compute</code> function into a data frame.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"SRR5947006.sig.gz"</span>, package<span class="op">=</span><span class="st">"sourmashconsumr"</span><span class="op">)</span></span>
<span><span class="va">signature_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_signature.html">read_signature</a></span><span class="op">(</span><span class="va">fpath</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">signature_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 16</span></span></span>
<span><span class="co">#&gt;   class    email hash_function filename name  license   num ksize  seed max_hash</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> sourmas… <span style="color: #949494;">""</span>    0.murmur64    SRR5947… SRR5… CC0         0    21    42  9.22<span style="color: #949494;">e</span>13</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> sourmas… <span style="color: #949494;">""</span>    0.murmur64    SRR5947… SRR5… CC0         0    21    42  9.22<span style="color: #949494;">e</span>13</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> sourmas… <span style="color: #949494;">""</span>    0.murmur64    SRR5947… SRR5… CC0         0    21    42  9.22<span style="color: #949494;">e</span>13</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> sourmas… <span style="color: #949494;">""</span>    0.murmur64    SRR5947… SRR5… CC0         0    21    42  9.22<span style="color: #949494;">e</span>13</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> sourmas… <span style="color: #949494;">""</span>    0.murmur64    SRR5947… SRR5… CC0         0    21    42  9.22<span style="color: #949494;">e</span>13</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> sourmas… <span style="color: #949494;">""</span>    0.murmur64    SRR5947… SRR5… CC0         0    21    42  9.22<span style="color: #949494;">e</span>13</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 6 more variables: scaled &lt;dbl&gt;, mins &lt;dbl&gt;, md5sum &lt;chr&gt;, abundances &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   molecule &lt;chr&gt;, version &lt;dbl&gt;</span></span></span></code></pre></div>
<p>If you have many signatures that you would like to read into the same
data frame, you can pass a character vector of file paths:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># grab the files paths for two signatures in the sourmashconsumr example data</span></span>
<span><span class="va">fpath1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"SRR5947006.sig.gz"</span>, package<span class="op">=</span><span class="st">"sourmashconsumr"</span><span class="op">)</span></span>
<span><span class="va">fpath2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"SRR5935765.sig.gz"</span>, package<span class="op">=</span><span class="st">"sourmashconsumr"</span><span class="op">)</span></span>
<span><span class="va">fpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">fpath1</span>, <span class="va">fpath2</span><span class="op">)</span></span>
<span><span class="co"># read in the signature files</span></span>
<span><span class="va">signature_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_signature.html">read_signature</a></span><span class="op">(</span><span class="va">fpaths</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="reading-the-csv-output-of-sourmash-compare">Reading the CSV output of sourmash compare<a class="anchor" aria-label="anchor" href="#reading-the-csv-output-of-sourmash-compare"></a>
</h4>
<p>The sourmash compare output is a square matrix. The CSV has column
names but no row names, but the rownames are symmetrical to the column
names. The <code><a href="../reference/read_compare_csv.html">read_compare_csv()</a></code> function reads in the compare
CSV file and adds row names. You can choose whether the row labels are
read in as a column named “sample” or as row names using the
<code>sample_to_rownames</code> argument.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"gut_compare.csv.gz"</span>, package <span class="op">=</span> <span class="st">"sourmashconsumr"</span><span class="op">)</span></span>
<span><span class="va">compare_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_compare_csv.html">read_compare_csv</a></span><span class="op">(</span><span class="va">fpath</span>, sample_to_rownames <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">6</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">6</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> ","</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (6): SRR5935765, SRR5936131, SRR5936197, SRR5946920, SRR5946923, SRR5947006</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span><span class="va">compare_df</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">#&gt;   sample     SRR5935765 SRR5936131 SRR5936197 SRR5946920 SRR5946923 SRR5947006</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> SRR5935765      1          0.283      0.269     0.139       0.197     0.115 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> SRR5936131      0.283      1          0.488     0.148       0.350     0.135 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> SRR5936197      0.269      0.488      1         0.183       0.546     0.134 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> SRR5946920      0.139      0.148      0.183     1           0.175     0.094<span style="text-decoration: underline;">3</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> SRR5946923      0.197      0.350      0.546     0.175       1         0.172 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> SRR5947006      0.115      0.135      0.134     0.094<span style="text-decoration: underline;">3</span>      0.172     1</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="reading-the-csv-output-of-sourmash-gather">Reading the CSV output of sourmash gather<a class="anchor" aria-label="anchor" href="#reading-the-csv-output-of-sourmash-gather"></a>
</h4>
<p>The <code><a href="../reference/read_gather.html">read_gather()</a></code> function reads in the CSV output by
sourmash gather. You can optionally filter the matches using
<code>intersect_bp_threshold</code>. The function does no filtering by
default. However, sourmash gather uses a threshold of 50,000 base pairs,
so if you decreased this value when you ran sourmash gather, it may make
sense to filter with <code>intersect_bp_thershold = 50000</code>. Most
false positives are caused by contamination in the reference genomes in
the database, so low filtering thresholds can improve accuracy.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"SRR5946920_gather_gtdbrs207_reps.csv.gz"</span>, package <span class="op">=</span> <span class="st">"sourmashconsumr"</span><span class="op">)</span></span>
<span><span class="va">gather_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_gather.html">read_gather</a></span><span class="op">(</span><span class="va">fpath</span>, intersect_bp_threshold <span class="op">=</span> <span class="fl">50000</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gather_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 30</span></span></span>
<span><span class="co">#&gt;   intersect_bp f_orig_query f_match f_unique_to_query f_unique_weighted</span></span>
<span><span class="co">#&gt;          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>      3<span style="text-decoration: underline;">787</span>000       0.031<span style="text-decoration: underline;">8</span>   0.583            0.031<span style="text-decoration: underline;">8</span>            0.049<span style="text-decoration: underline;">6</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>      3<span style="text-decoration: underline;">544</span>000       0.029<span style="text-decoration: underline;">8</span>   0.707            0.029<span style="text-decoration: underline;">7</span>            0.095<span style="text-decoration: underline;">6</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>      3<span style="text-decoration: underline;">622</span>000       0.030<span style="text-decoration: underline;">4</span>   0.773            0.029<span style="text-decoration: underline;">4</span>            0.029<span style="text-decoration: underline;">8</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>      2<span style="text-decoration: underline;">993</span>000       0.025<span style="text-decoration: underline;">2</span>   0.613            0.025<span style="text-decoration: underline;">1</span>            0.014<span style="text-decoration: underline;">0</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>      2<span style="text-decoration: underline;">941</span>000       0.024<span style="text-decoration: underline;">7</span>   0.619            0.024<span style="text-decoration: underline;">0</span>            0.053<span style="text-decoration: underline;">1</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>      2<span style="text-decoration: underline;">389</span>000       0.020<span style="text-decoration: underline;">1</span>   0.585            0.019<span style="text-decoration: underline;">2</span>            0.062<span style="text-decoration: underline;">4</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 25 more variables: average_abund &lt;dbl&gt;, median_abund &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   std_abund &lt;dbl&gt;, filename &lt;chr&gt;, name &lt;chr&gt;, genome_accession &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   md5 &lt;chr&gt;, f_match_orig &lt;dbl&gt;, unique_intersect_bp &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   gather_result_rank &lt;dbl&gt;, remaining_bp &lt;dbl&gt;, query_filename &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   query_name &lt;chr&gt;, query_md5 &lt;chr&gt;, query_bp &lt;dbl&gt;, ksize &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   moltype &lt;chr&gt;, scaled &lt;dbl&gt;, query_n_hashes &lt;dbl&gt;, query_abundance &lt;lgl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   query_containment_ani &lt;dbl&gt;, match_containment_ani &lt;dbl&gt;, …</span></span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="reading-the-csv-output-of-sourmash-taxonomy-annotate">Reading the CSV output of sourmash taxonomy annotate<a class="anchor" aria-label="anchor" href="#reading-the-csv-output-of-sourmash-taxonomy-annotate"></a>
</h4>
<p>The <code><a href="../reference/read_taxonomy_annotate.html">read_taxonomy_annotate()</a></code> function reads the CSV
output by sourmash taxonomy annotate. Taxonomy annotate adds a taxonomic
lineage column to the sourmash gather output. You can filter with
<code>intersect_bp_threshold</code> (see above). You can also separate
the lineage into taxonomy ranks using <code>separate_lineage</code>.
This function assumes that the lineage ranks are separated by
semi-colons (;), and that the lineage is composed of a domain, phylum,
class, order, family, genus, species, and strain. If information is
missing, it is back-filled with <code>NA</code>s.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"SRR5946920_gather_gtdbrs207_reps.with-lineages.csv.gz"</span>, package <span class="op">=</span> <span class="st">"sourmashconsumr"</span><span class="op">)</span></span>
<span><span class="va">taxonomy_annotate_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_taxonomy_annotate.html">read_taxonomy_annotate</a></span><span class="op">(</span><span class="va">fpath</span>, </span>
<span>                                               intersect_bp_threshold <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                                               separate_lineage <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">taxonomy_annotate_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 40</span></span></span>
<span><span class="co">#&gt;   intersect_bp f_orig_query f_match f_unique_to_query f_unique_weighted</span></span>
<span><span class="co">#&gt;          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>      3<span style="text-decoration: underline;">787</span>000       0.031<span style="text-decoration: underline;">8</span>   0.583            0.031<span style="text-decoration: underline;">8</span>            0.049<span style="text-decoration: underline;">6</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>      3<span style="text-decoration: underline;">544</span>000       0.029<span style="text-decoration: underline;">8</span>   0.707            0.029<span style="text-decoration: underline;">7</span>            0.095<span style="text-decoration: underline;">6</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>      3<span style="text-decoration: underline;">622</span>000       0.030<span style="text-decoration: underline;">4</span>   0.773            0.029<span style="text-decoration: underline;">4</span>            0.029<span style="text-decoration: underline;">8</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>      2<span style="text-decoration: underline;">993</span>000       0.025<span style="text-decoration: underline;">2</span>   0.613            0.025<span style="text-decoration: underline;">1</span>            0.014<span style="text-decoration: underline;">0</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>      2<span style="text-decoration: underline;">941</span>000       0.024<span style="text-decoration: underline;">7</span>   0.619            0.024<span style="text-decoration: underline;">0</span>            0.053<span style="text-decoration: underline;">1</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span>      2<span style="text-decoration: underline;">389</span>000       0.020<span style="text-decoration: underline;">1</span>   0.585            0.019<span style="text-decoration: underline;">2</span>            0.062<span style="text-decoration: underline;">4</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 35 more variables: average_abund &lt;dbl&gt;, median_abund &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   std_abund &lt;dbl&gt;, filename &lt;chr&gt;, name &lt;chr&gt;, genome_accession &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   md5 &lt;chr&gt;, f_match_orig &lt;dbl&gt;, unique_intersect_bp &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   gather_result_rank &lt;dbl&gt;, remaining_bp &lt;dbl&gt;, query_filename &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   query_name &lt;chr&gt;, query_md5 &lt;chr&gt;, query_bp &lt;dbl&gt;, ksize &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   moltype &lt;chr&gt;, scaled &lt;dbl&gt;, query_n_hashes &lt;dbl&gt;, query_abundance &lt;lgl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   query_containment_ani &lt;dbl&gt;, match_containment_ani &lt;dbl&gt;, …</span></span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="working-with-sourmash-signatures-fracminhash-sketches">Working with sourmash signatures (FracMinHash sketches)<a class="anchor" aria-label="anchor" href="#working-with-sourmash-signatures-fracminhash-sketches"></a>
</h3>
<p>Sourmash signatures are compressed representations of sequencing
data. This data can be used for many different purposes such as machine
learning<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="https://www.biorxiv.org/content/10.1101/2022.06.30.498290v1" class="external-link uri"&gt;https://www.biorxiv.org/content/10.1101/2022.06.30.498290v1&lt;/a&gt;&lt;/p&gt;'><sup>4</sup></a>, differential k-mer abundance analysis, and
sequencing depth/rarefaction analysis.</p>
<div class="section level4">
<h4 id="visualizing-shared-k-mer-content-between-samples-with-an-upset-plot">Visualizing shared k-mer content between samples with an upset
plot<a class="anchor" aria-label="anchor" href="#visualizing-shared-k-mer-content-between-samples-with-an-upset-plot"></a>
</h4>
<p>The functions <code><a href="../reference/from_signatures_to_upset_df.html">from_signatures_to_upset_df()</a></code> and
<code><a href="../reference/plot_signatures_upset.html">plot_signatures_upset()</a></code> produce an upset plot that depicts
the number of k-mers shared between sets of samples. The absolute size
of the intersection is the number of shared hashes (k-mers) in scaled
FracMinHash sketches (see <a href="#how-sourmash-works">How sourmash
works</a> for an explanation of sketches and their content.). These
plots do not take abundance information into account.</p>
<p>An upset plot is an alternative to a venn diagram; the bottom half of
the plot shows which samples have shared content while the top half of
the plot shows the size of the intersection between the specified
samples.</p>
<p>This plot is good for visualizing the exact sequencing content
overlap between samples. If there is a lot of variation in the samples,
it might be difficult to interpret the plot as there may be many
intersections. Still, it provides a nice complement to the sourmash
compare results and plots as it gives more information about the nature
of the intersections between samples.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gut_signatures_df</span> <span class="op">&lt;-</span> <span class="va">gut_signatures_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">ksize</span> <span class="op">==</span> <span class="fl">31</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gut_signatures_upset_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/from_signatures_to_upset_df.html">from_signatures_to_upset_df</a></span><span class="op">(</span>signatures_df <span class="op">=</span> <span class="va">gut_signatures_df</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gut_signatures_upset_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;              SRR5935765 SRR5936131 SRR5936197 SRR5946920 SRR5946923 SRR5947006</span></span>
<span><span class="co">#&gt; 7504968134            1          1          0          1          0          1</span></span>
<span><span class="co">#&gt; 47418666416           1          0          0          0          0          0</span></span>
<span><span class="co">#&gt; 91683251943           1          1          0          1          0          1</span></span>
<span><span class="co">#&gt; 133675644697          1          0          0          0          0          0</span></span>
<span><span class="co">#&gt; 232110850204          1          0          0          0          0          0</span></span>
<span><span class="co">#&gt; 320454190342          1          1          1          1          1          0</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_signatures_upset.html">plot_signatures_upset</a></span><span class="op">(</span><span class="va">gut_signatures_upset_df</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-10-1.png" width="768"></p>
<p>Any additional arguments that can be passed to
<code><a href="https://krassowski.github.io/complex-upset/reference/upset.html" class="external-link">ComplexUpset::upset()</a></code> can be added to the end of the
function to further control the look of the plot. <a href="https://krassowski.github.io/complex-upset/articles/Examples_R.html" class="external-link">ComplexUpset</a>
has a lot of documentation and examples for modifying plots. Below, we
use some of these arguments to modify common parts of the upset plot. We
use bright colors to make it obvious which components of the plot the
arguments modify.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_signatures_upset.html">plot_signatures_upset</a></span><span class="op">(</span>upset_df <span class="op">=</span> <span class="va">gut_signatures_upset_df</span>,</span>
<span>                      <span class="co"># arguments to ComplexUpset::upset():</span></span>
<span>                      stripes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cyan"</span><span class="op">)</span>, </span>
<span>                      name <span class="op">=</span> <span class="st">"Groups!"</span>, </span>
<span>                      min_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                      themes <span class="op">=</span> <span class="fu">ComplexUpset</span><span class="fu">::</span><span class="fu"><a href="https://krassowski.github.io/complex-upset/reference/upset_default_themes.html" class="external-link">upset_default_themes</a></span><span class="op">(</span>text<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">16</span>, color <span class="op">=</span> <span class="st">"orange"</span><span class="op">)</span>,</span>
<span>                                                                  axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">13</span>, color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span>,</span>
<span>                                                                  panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="assessing-sample-sequencing-depth-using-rarefaction-accumulation-curves">Assessing sample sequencing depth using rarefaction (accumulation)
curves<a class="anchor" aria-label="anchor" href="#assessing-sample-sequencing-depth-using-rarefaction-accumulation-curves"></a>
</h4>
<p>The goal of the <code><a href="../reference/from_signatures_to_rarefaction_df.html">from_signatures_to_rarefaction_df()</a></code> and
<code><a href="../reference/plot_signatures_rarefaction.html">plot_signatures_rarefaction()</a></code> functions is to produce a
rarefaction curve to assess whether an adequate sequencing depth was
reached to capture the complexity of the sample sequenced. These
functions should work on raw or trimmed reads from short or long read
sequencing data generated from most library types (e.g. genomes,
transcriptomes, metagenomes, metatranscriptomes, etc.). If your library
is non-complex (e.g. amplicon sequencing data), you may need to use a
lower scaled value. Rarefaction curves rely on the abundance of
sequences in a sample. Therefore, it only makes sense to generate
rarefaction curves from signatures that were made with FASTQ files and
that had abundance tracking turned on. Further, these curves may be more
accurate if you build them with sequences from many samples from similar
environments. In case signatures were built from reads that have not
been k-mer trimmed, there is a filtering step that removes hashes
(k-mers) that are only observed once and only observed in one sample as
these are likely sequencing errors. Please note that this filtering
process may invalidate downstream rarefaction curve convergence
estimation as many of these methods evaluate singletons and doubletons
in the data set.</p>
<p>The function <code><a href="../reference/from_signatures_to_rarefaction_df.html">from_signatures_to_rarefaction_df()</a></code> may
take a while to run. We’ve found it produces fairly consistent results
even when signatures are dramatically downsampled with a high scaled
value (e.g. <code>scaled = 100000</code>). You can downsample your
signatures with sourmash prior to reading them into R using the function
<code>sourmash sig downsample</code>. See the <a href="https://sourmash.readthedocs.io" class="external-link">sourmash documentation</a> for
more details.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filter to a single k-mer size</span></span>
<span><span class="va">gut_signatures_df</span> <span class="op">&lt;-</span> <span class="va">gut_signatures_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">ksize</span> <span class="op">==</span> <span class="fl">31</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rarefaction_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/from_signatures_to_rarefaction_df.html">from_signatures_to_rarefaction_df</a></span><span class="op">(</span>signatures_df <span class="op">=</span> <span class="va">gut_signatures_df</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rarefaction_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;         name num_kmers_sampled num_kmers_observed</span></span>
<span><span class="co">#&gt; 1 SRR5935765                 1           1.000000</span></span>
<span><span class="co">#&gt; 2 SRR5935765                 2           1.993401</span></span>
<span><span class="co">#&gt; 3 SRR5935765                 3           2.980279</span></span>
<span><span class="co">#&gt; 4 SRR5935765                 4           3.960713</span></span>
<span><span class="co">#&gt; 5 SRR5935765                 5           4.934776</span></span>
<span><span class="co">#&gt; 6 SRR5935765                 6           5.902544</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rarefaction_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_signatures_rarefaction.html">plot_signatures_rarefaction</a></span><span class="op">(</span>rarefaction_df <span class="op">=</span> <span class="va">rarefaction_df</span>, fraction_of_points_to_plot <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">rarefaction_plt</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/plot_signatures-rarefaction-1.png" width="700"></p>
<p>The plot produced by <code>plot_signatures_rarefaction</code> is a
ggplot2 plot, so you can modify it by adding additional layers:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rarefaction_plt</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/plot_signatures_rarefaction_mod-1.png" width="700"></p>
<p>Color by sample name:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rarefaction_plt</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">name</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/plot_signatures_rarefaction_mod2-1.png" width="700"></p>
<p>Add your own metadata to the <code>rarefaction_df</code> and use it
in the plot:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># join the metadata data frame with the rarefaction_df</span></span>
<span><span class="va">rarefaction_df</span> <span class="op">&lt;-</span> <span class="va">rarefaction_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">metadata</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"name"</span> <span class="op">=</span> <span class="st">"run_accessions"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create and modify a plot</span></span>
<span><span class="fu"><a href="../reference/plot_signatures_rarefaction.html">plot_signatures_rarefaction</a></span><span class="op">(</span>rarefaction_df <span class="op">=</span> <span class="va">rarefaction_df</span>, fraction_of_points_to_plot <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">groups</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/plot_signatures_rarefaction_mod3-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="working-with-the-csv-output-of-sourmash-compare">Working with the CSV output of sourmash compare<a class="anchor" aria-label="anchor" href="#working-with-the-csv-output-of-sourmash-compare"></a>
</h3>
<p>Sourmash compare outputs an all-by-all similarity matrix for a group
of samples (by default, the matrix is a similarity matrix, but users can
switch this to a dissimilarity matrix with a command line flag). When
the samples were sketched with abundance tracking as was done here
(e.g. the abundance of each k-mer was recorded), the similarity matrix
encodes the angular distance. The sourmashconsumr package supports two
visualizations from the CSV output of sourmash compare, an MDS
(ordination) plot and a clustered heat map.</p>
<div class="section level4">
<h4 id="mds-plot-from-sourmash-compare-output">MDS plot from sourmash compare output<a class="anchor" aria-label="anchor" href="#mds-plot-from-sourmash-compare-output"></a>
</h4>
<p>The MDS plot allows you to visualize how similar each sample is to
every other sample in two dimensional space. Samples that cluster more
closely together are most similar.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gut_compare_mds_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_compare_mds.html">make_compare_mds</a></span><span class="op">(</span><span class="va">gut_compare_df</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gut_compare_mds_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;       sample       MDS1        MDS2</span></span>
<span><span class="co">#&gt; 1 SRR5935765  0.0775339  0.13464650</span></span>
<span><span class="co">#&gt; 2 SRR5936131 -0.3745227 -0.03349708</span></span>
<span><span class="co">#&gt; 3 SRR5936197 -0.5022308 -0.03442290</span></span>
<span><span class="co">#&gt; 4 SRR5946920  0.5039572  0.64805445</span></span>
<span><span class="co">#&gt; 5 SRR5946923 -0.3743700 -0.10706435</span></span>
<span><span class="co">#&gt; 6 SRR5947006  0.6696324 -0.60771661</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compare_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_compare_mds.html">plot_compare_mds</a></span><span class="op">(</span><span class="va">gut_compare_mds_df</span><span class="op">)</span></span>
<span><span class="va">compare_plt</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>The MDS plot is a ggplot2 object, so you can modify it by adding
layers.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">compare_plt</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>You can also add metadata directly to the data frame output by
<code>make_compare_mds</code> so that you can add it to a plot:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gut_compare_mds_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">metadata</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample"</span> <span class="op">=</span> <span class="st">"run_accessions"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;       sample       MDS1        MDS2 groups</span></span>
<span><span class="co">#&gt; 1 SRR5935765  0.0775339  0.13464650     cd</span></span>
<span><span class="co">#&gt; 2 SRR5936131 -0.3745227 -0.03349708     cd</span></span>
<span><span class="co">#&gt; 3 SRR5936197 -0.5022308 -0.03442290 nonibd</span></span>
<span><span class="co">#&gt; 4 SRR5946920  0.5039572  0.64805445 nonibd</span></span>
<span><span class="co">#&gt; 5 SRR5946923 -0.3743700 -0.10706435 nonibd</span></span>
<span><span class="co">#&gt; 6 SRR5947006  0.6696324 -0.60771661     cd</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_compare_mds.html">plot_compare_mds</a></span><span class="op">(</span><span class="va">gut_compare_mds_df</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">groups</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="heatmap-from-sourmash-compare-output">Heatmap from sourmash compare output<a class="anchor" aria-label="anchor" href="#heatmap-from-sourmash-compare-output"></a>
</h4>
<p>The <code>plot_compare_heatmap</code> object works directly on the
output of <code><a href="../reference/read_compare_csv.html">read_compare_csv()</a></code>. It’s a wrapper function for
<code>base::heatmap()</code>, so you can add any arguments accepted by
<code>base::heatmap()</code> to modify the look of the plot.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_compare_heatmap.html">plot_compare_heatmap</a></span><span class="op">(</span><span class="va">gut_compare_df</span>, cexRow <span class="op">=</span> <span class="fl">0.75</span>, cexCol <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-16-1.png" width="576"></p>
</div>
</div>
<div class="section level3">
<h3 id="working-with-the-csv-output-of-sourmash-gather">Working with the CSV output of sourmash gather<a class="anchor" aria-label="anchor" href="#working-with-the-csv-output-of-sourmash-gather"></a>
</h3>
<p>Sourmash gather compares a query (here, a metagenome) against a
database (here, GTDB rs207 representatives) and provides the minimum set
of genomes in the database that cover (contain) all of the k-mers in the
query.</p>
<p>Sourmash gather works best for metagenomes from environments that
have been sequenced before or from which many genomes have been isolated
and sequenced. Because k-mers are so specific, a genome needs to be in a
database for sourmash gather to find a match. Sourmash gather won’t find
much above species (k = 31) or genus (k = 21) similarity, so if most of
the organisms in a sample are new, sourmash won’t be able to label
them.</p>
<p>Sourmashconsumr produces two plots from the sourmash gather results:
a high-level bargraph summarizes the fraction of the sample that had
matches in the database, and an upset plot showing intersections between
genomes that were matched in multiple samples. These plots are designed
to visualize the output of running sourmash gather on many samples in
the same plot.</p>
<div class="section level4">
<h4 id="fraction-classifed-plot-from-sourmash-gather-output">Fraction classifed plot from sourmash gather output<a class="anchor" aria-label="anchor" href="#fraction-classifed-plot-from-sourmash-gather-output"></a>
</h4>
<p>The actual number that is plotted is referred to as
abundance-weighted unique fraction. By default, sourmash gather returns
the following information:</p>
<ul>
<li>
<code>f_orig_query</code>: the fraction of the original query
(metagenome sample) that matched against the genome</li>
<li>
<code>f_unique_to_query</code>: the fraction of the matched genome
that was unique to the query. Given that multiple matched genomes could
cover the same portion of the query, this column reports the fraction of
the query that was covered by the a matched genome, but doesn’t allow
multiple genomes to cover the same portion. As such, the genome that
covers the most k-mers in a metagenome sample gets to “anchor” those
k-mers, while other genomes that could cover those k-mers don’t get any
of them. This column removes the double counting problem that arises
from having multiple genomes in a database that contain some of the same
sequences.</li>
<li>
<code>f_unique_weighted</code>: This column weights the
f_unique_to_query by the abundance of those k-mers. This is important in
the context of a metagenome, as the abundance reflects the abundance of
the genome in the original microbial community.</li>
</ul>
<p>The plots below use <code>f_unique_weighted</code> to look at the
taxonomic composition of each metagenome sample. This value is similar
to looking at the fraction of reads that would map against reference
genomes.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_gather_classified.html">plot_gather_classified</a></span><span class="op">(</span>gather_df <span class="op">=</span> <span class="va">gut_gather_df</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="upset-plot-from-sourmash-gather-output">Upset plot from sourmash gather output<a class="anchor" aria-label="anchor" href="#upset-plot-from-sourmash-gather-output"></a>
</h4>
<p>An upset plot is an alternative to a venn diagram; the bottom half of
the plot shows which samples have shared content while the top half of
the plot shows the size of the intersection between the specified
samples.</p>
<p>This upset plot visualizes the intersection of genomes that were
identified in different samples.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gut_gather_upset_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/from_gather_to_upset_df.html">from_gather_to_upset_df</a></span><span class="op">(</span>gather_df <span class="op">=</span> <span class="va">gut_gather_df</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gut_gather_upset_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 SRR5935765 SRR5936131 SRR5936197 SRR5946920 SRR5946923</span></span>
<span><span class="co">#&gt; GCF_002234575.2          1          1          0          1          0</span></span>
<span><span class="co">#&gt; GCF_001314995.1          1          1          1          1          1</span></span>
<span><span class="co">#&gt; GCF_000154205.1          1          1          1          1          1</span></span>
<span><span class="co">#&gt; GCF_000012825.1          1          1          1          1          1</span></span>
<span><span class="co">#&gt; GCF_002222615.2          1          1          1          1          1</span></span>
<span><span class="co">#&gt; GCF_000466485.1          1          1          0          1          0</span></span>
<span><span class="co">#&gt;                 SRR5947006</span></span>
<span><span class="co">#&gt; GCF_002234575.2          1</span></span>
<span><span class="co">#&gt; GCF_001314995.1          1</span></span>
<span><span class="co">#&gt; GCF_000154205.1          1</span></span>
<span><span class="co">#&gt; GCF_000012825.1          1</span></span>
<span><span class="co">#&gt; GCF_002222615.2          0</span></span>
<span><span class="co">#&gt; GCF_000466485.1          1</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_gather_upset.html">plot_gather_upset</a></span><span class="op">(</span>upset_df <span class="op">=</span> <span class="va">gut_gather_upset_df</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-19-1.png" width="768"></p>
<p>The bars in the upset plot can optionally be colored by the database
the match came from. This is useful when you use many databases when
running sourmash gather. Since we only used one database to classify our
samples in the example data, the visualization isn’t much more
rewarding.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_gather_upset.html">plot_gather_upset</a></span><span class="op">(</span>upset_df <span class="op">=</span> <span class="va">gut_gather_upset_df</span>, </span>
<span>                  color_by_database <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                  gather_df <span class="op">=</span> <span class="va">gut_gather_df</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-20-1.png" width="1056"></p>
<p>You can change the palette by passing a character vector of colors to
the <code>palette</code> argument. If no palette is specified, it
defaults to RColorBrewer’s Set2.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_gather_upset.html">plot_gather_upset</a></span><span class="op">(</span>upset_df <span class="op">=</span> <span class="va">gut_gather_upset_df</span>, </span>
<span>                  color_by_database <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                  gather_df <span class="op">=</span> <span class="va">gut_gather_df</span>,</span>
<span>                  palette <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-21-1.png" width="1056"></p>
<p>See the <a href="#visualizing-shared-k-mer-content-between-samples-with-an-upset-plot">Visualizing
shared k-mer content between samples with an upset plot</a> for examples
of how to modify the look of the upset plot.</p>
</div>
</div>
<div class="section level3">
<h3 id="working-with-the-csv-output-of-sourmash-taxonomy-annotate">Working with the CSV output of sourmash taxonomy annotate<a class="anchor" aria-label="anchor" href="#working-with-the-csv-output-of-sourmash-taxonomy-annotate"></a>
</h3>
<p>Sourmash taxonomy makes the sourmash gather output more interpretable
by adding taxonomic labels to the sourmash gather results. Sourmash
gather only outputs the statistics about the samples in a database that
were found in a query sample. Taxonomic labels contextualize these
results and allow new analyses like agglomeration up the taxonomic
lineage to summarize results or visualizing results in the context of
taxonomy.</p>
<div class="section level4">
<h4 id="agglomerating-up-levels-of-taxonomy">Agglomerating up levels of taxonomy<a class="anchor" aria-label="anchor" href="#agglomerating-up-levels-of-taxonomy"></a>
</h4>
<p>Inspired by <code><a href="https://rdrr.io/pkg/phyloseq/man/tax_glom.html" class="external-link">phyloseq::tax_glom()</a></code>, this method summarizes
variables output by sourmash gather from genomes that have the same
taxonomy at a user-specified taxonomy rank. Agglomeration occurs within
each sample, meaning the variable is only summed within each query_name.
This function returns a data frame with the columns ‘lineage’,
‘query_name’, and the variable that is specified by
<code>glom_var</code>.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">glom_df</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/tax_glom_taxonomy_annotate.html">tax_glom_taxonomy_annotate</a></span><span class="op">(</span>taxonomy_annotate_df <span class="op">=</span> <span class="va">gut_taxonomy_annotate_df</span>,</span>
<span>                                       tax_glom_level <span class="op">=</span> <span class="st">"genus"</span>,</span>
<span>                                       glom_var <span class="op">=</span> <span class="st">"f_unique_to_query"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">glom_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt;   lineage                                           query_name f_unique_to_query</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> d__Bacteria;p__Actinobacteriota;c__Actinomycetia… SRR5935765          0.001<span style="text-decoration: underline;">32</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Ba… SRR5935765          0.000<span style="text-decoration: underline;">956</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Ba… SRR5935765          0.133   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Ba… SRR5935765          0.052<span style="text-decoration: underline;">5</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Ba… SRR5935765          0.001<span style="text-decoration: underline;">68</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Ba… SRR5935765          0.000<span style="text-decoration: underline;">856</span></span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visualizing-shared-taxonomic-lineages-between-samples-with-an-upset-plot">Visualizing shared taxonomic lineages between samples with an upset
plot<a class="anchor" aria-label="anchor" href="#visualizing-shared-taxonomic-lineages-between-samples-with-an-upset-plot"></a>
</h4>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gut_taxonomy_upset_inputs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/from_taxonomy_annotate_to_upset_inputs.html">from_taxonomy_annotate_to_upset_inputs</a></span><span class="op">(</span>taxonomy_annotate_df <span class="op">=</span> <span class="va">gut_taxonomy_annotate_df</span>, </span>
<span>                                                                    tax_glom_level <span class="op">=</span> <span class="st">"order"</span><span class="op">)</span></span></code></pre></div>
<p>This function produces a list with three objects. The first,
<code>gut_taxonomy_upset_inputs$upset_df</code> is an upset-formatted
data frame indicating the presense/absence of each lineage for each
sample. The second,
<code>gut_taxonomy_upset_inputs$taxonomy_annotate_df</code>, is a
re-formatted data frame with the taxonomy annotation information. The
third, <code>gut_taxonomy_upset_inputs$tax_glom_level</code>, records
the taxonomy agglomeration level used to create the upset inputs.</p>
<p>This list can then be fed into
<code><a href="../reference/plot_taxonomy_annotate_upset.html">plot_taxonomy_annotate_upset()</a></code> to produce an upset
plot:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_taxonomy_annotate_upset.html">plot_taxonomy_annotate_upset</a></span><span class="op">(</span>upset_inputs <span class="op">=</span> <span class="va">gut_taxonomy_upset_inputs</span>, </span>
<span>                             fill <span class="op">=</span> <span class="st">"phylum"</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-24-1.png" width="576"></p>
<p>If fill is specified, the bar plot on the upper portion of the upset
plot will be colored by the taxonomic level indicated. The level must be
at or above that used in
<code><a href="../reference/from_taxonomy_annotate_to_upset_inputs.html">from_taxonomy_annotate_to_upset_inputs()</a></code>.</p>
<p>You can change the palette by passing a character vector of colors to
the <code>palette</code> argument. If no palette is specified, it
defaults to RColorBrewer’s Set2.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_taxonomy_annotate_upset.html">plot_taxonomy_annotate_upset</a></span><span class="op">(</span>upset_inputs <span class="op">=</span> <span class="va">gut_taxonomy_upset_inputs</span>, </span>
<span>                             fill <span class="op">=</span> <span class="st">"phylum"</span>,</span>
<span>                             palette <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#999999"</span>, <span class="st">"#E69F00"</span>, <span class="st">"#56B4E9"</span>, <span class="st">"#009E73"</span>,</span>
<span>                                         <span class="st">"#F0E442"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#D55E00"</span>, <span class="st">"#CC79A7"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-25-1.png" width="576"></p>
<p>Since upset plots are based on the presence or absence of lineages,
the counts in the upset barplot do not reflect the total number of
lineages observed at a given taxonomic level, but rather how many
lineages are shared between samples at that level – e.g., the numbers
are not weighted.</p>
<p>See the <a href="#visualizing-shared-k-mer-content-between-samples-with-an-upset-plot">Visualizing
shared k-mer content between samples with an upset plot</a> for examples
of how to modify the look of the upset plot.</p>
</div>
<div class="section level4">
<h4 id="visualizing-taxonomy-with-a-sankey-plot">Visualizing taxonomy with a Sankey plot<a class="anchor" aria-label="anchor" href="#visualizing-taxonomy-with-a-sankey-plot"></a>
</h4>
<p>The Sankey plot is designed to show the taxonomic breakdown in one or
many samples. The ribbons in the plot progressively split as the
taxonomic lineage becomes more specific (e.g. from domain to phylum,
phylum to class, etc.). The ribbon width displays the relative abundance
of the taxonomic lineage. The <code>tax_glom_level</code> specifies
which level of the taxonomic lineage to stop displaying the
taxonomy.</p>
<p>The Sankey plot can be generated from one or many samples. Below is
an example of a plot generated from one sample:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_taxonomy_annotate_sankey.html">plot_taxonomy_annotate_sankey</a></span><span class="op">(</span>taxonomy_annotate_df <span class="op">=</span> <span class="va">gut_taxonomy_annotate_df</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>                                <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">query_name</span> <span class="op">==</span> <span class="st">"SRR5935765"</span><span class="op">)</span>, </span>
<span>                              tax_glom_level <span class="op">=</span> <span class="st">"family"</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-26-1.png" width="768"></p>
<p>And a plot generated from many samples:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_taxonomy_annotate_sankey.html">plot_taxonomy_annotate_sankey</a></span><span class="op">(</span>taxonomy_annotate_df <span class="op">=</span> <span class="va">gut_taxonomy_annotate_df</span>, </span>
<span>                              tax_glom_level <span class="op">=</span> <span class="st">"order"</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-27-1.png" width="768"></p>
<p>The colors in the Sankey plot are built using the function
<code><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">grDevices::colorRampPalette()</a></code> on the RColorBrewer palette
Set2. To change the palette, supply the <code>palette</code> argument
with the hexcodes you would like to ramp up.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_taxonomy_annotate_sankey.html">plot_taxonomy_annotate_sankey</a></span><span class="op">(</span>taxonomy_annotate_df <span class="op">=</span> <span class="va">gut_taxonomy_annotate_df</span>, </span>
<span>                              tax_glom_level <span class="op">=</span> <span class="st">"order"</span>,</span>
<span>                              palette <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">8</span>, <span class="st">"Set3"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-28-1.png" width="768"></p>
<p>By default, taxonomic labels are displayed on the plot
(<code>label = TRUE</code>). If you would like to control the look of
the labels, use <code>label = FALSE</code> and then add a layer to the
output plot for the labels:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_taxonomy_annotate_sankey.html">plot_taxonomy_annotate_sankey</a></span><span class="op">(</span>taxonomy_annotate_df <span class="op">=</span> <span class="va">gut_taxonomy_annotate_df</span>, </span>
<span>                              tax_glom_level <span class="op">=</span> <span class="st">"order"</span>,</span>
<span>                              palette <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">8</span>, <span class="st">"Set3"</span><span class="op">)</span>,</span>
<span>                              label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggforce</span><span class="fu">::</span><span class="fu"><a href="https://ggforce.data-imaginist.com/reference/geom_parallel_sets.html" class="external-link">geom_parallel_sets_labels</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">'black'</span>, angle <span class="op">=</span> <span class="fl">360</span>, size <span class="op">=</span> <span class="fl">2</span>, hjust <span class="op">=</span> <span class="op">-</span><span class="fl">0.25</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-29-1.png" width="768"></p>
</div>
<div class="section level4">
<h4 id="visualizing-taxonomic-changes-over-time-with-a-time-series-alluvial-plot">Visualizing taxonomic changes over time with a time series alluvial
plot<a class="anchor" aria-label="anchor" href="#visualizing-taxonomic-changes-over-time-with-a-time-series-alluvial-plot"></a>
</h4>
<p>The time series alluvial plot shows how lineages observed in samples
change. It was designed to show how the fraction of taxonomic lineages
in community change over time when samples are taken in time series. One
level of taxonomy (e.g. order) is shown throughout the plot, and the
alluvial ribbons show how the fraction of those lineages change between
samples.</p>
<p>Our example data aren’t actually time series, so this plot doesn’t
really make sense. However, for demonstration purposes, we’ll use our
test data anyway to show how this plot would work if we were working
with time series. Eventually, we’ll update the vignette to show what
this looks like with real time series data.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">time_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>query_name <span class="op">=</span> <span class="va">metadata</span><span class="op">$</span><span class="va">run_accessions</span>,</span>
<span>                      time <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_taxonomy_annotate_ts_alluvial.html">plot_taxonomy_annotate_ts_alluvial</a></span><span class="op">(</span>taxonomy_annotate_df <span class="op">=</span> <span class="va">gut_taxonomy_annotate_df</span>,</span>
<span>                                   time_df <span class="op">=</span> <span class="va">time_df</span>, </span>
<span>                                   tax_glom_level <span class="op">=</span> <span class="st">"family"</span>,</span>
<span>                                   fraction_threshold <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-30-1.png" width="672"></p>
<p>In this plot, the <code>f_unique_weighted</code> value from the
sourmash gather results is agglomerated to the user-specified
<code>tax_glom_level</code>. The <code>fraction_threshold</code> is the
minimum fraction that a lineage needs to be present in at least one
sample in the time series for the lineage to get its own alluvial ribbon
on the plot. If lineage is present at less than this threshold, it will
get grouped into the “other” category.</p>
<p>The plot outputs by <code><a href="../reference/plot_taxonomy_annotate_ts_alluvial.html">plot_taxonomy_annotate_ts_alluvial()</a></code>
is a ggplot. To change the palette, add a layer to the plot.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_taxonomy_annotate_ts_alluvial.html">plot_taxonomy_annotate_ts_alluvial</a></span><span class="op">(</span>taxonomy_annotate_df <span class="op">=</span> <span class="va">gut_taxonomy_annotate_df</span>,</span>
<span>                                   time_df <span class="op">=</span> <span class="va">time_df</span>, </span>
<span>                                   tax_glom_level <span class="op">=</span> <span class="st">"family"</span>,</span>
<span>                                   fraction_threshold <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Paired"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-31-1.png" width="672"></p>
<p>Alternatively, you can set the palette by providing a character
vector to the <code>palette</code> argument:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_taxonomy_annotate_ts_alluvial.html">plot_taxonomy_annotate_ts_alluvial</a></span><span class="op">(</span>taxonomy_annotate_df <span class="op">=</span> <span class="va">gut_taxonomy_annotate_df</span>,</span>
<span>                                   time_df <span class="op">=</span> <span class="va">time_df</span>, </span>
<span>                                   tax_glom_level <span class="op">=</span> <span class="st">"family"</span>,</span>
<span>                                   fraction_threshold <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                                   palette <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'#A6CEE3'</span>, <span class="st">'#1F78B4'</span>, <span class="st">'#B2DF8A'</span>, </span>
<span>                                               <span class="st">'#33A02C'</span>, <span class="st">'#FB9A99'</span>, <span class="st">'#E31A1C'</span>, </span>
<span>                                               <span class="st">'#FDBF6F'</span>, <span class="st">'#FF7F00'</span>, <span class="st">'#CAB2D6'</span>, </span>
<span>                                               <span class="st">'#6A3D9A'</span>, <span class="st">'#FFFF99'</span>, <span class="st">'#B15928'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-32-1.png" width="672"></p>
<p>By default, taxonomic labels are displayed on the plot
(<code>label = TRUE</code>). If you would like to control the look of
the labels, use <code>label = FALSE</code> and then add a layer to the
output plot for the labels:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_taxonomy_annotate_ts_alluvial.html">plot_taxonomy_annotate_ts_alluvial</a></span><span class="op">(</span>taxonomy_annotate_df <span class="op">=</span> <span class="va">gut_taxonomy_annotate_df</span>,</span>
<span>                                   time_df <span class="op">=</span> <span class="va">time_df</span>, </span>
<span>                                   tax_glom_level <span class="op">=</span> <span class="st">"family"</span>,</span>
<span>                                   fraction_threshold <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                                   label <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggalluvial</span><span class="fu">::</span><span class="fu"><a href="http://corybrunson.github.io/ggalluvial/reference/stat_alluvium.html" class="external-link">stat_alluvium</a></span><span class="op">(</span>geom <span class="op">=</span> <span class="st">"text"</span>, size <span class="op">=</span> <span class="fl">2</span>, decreasing <span class="op">=</span> <span class="cn">FALSE</span>, min.y <span class="op">=</span> <span class="fl">0.005</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-33-1.png" width="672"></p>
</div>
<div class="section level4">
<h4 id="converting-from-taxonomy-annotate-to-phyloseq-object">Converting from taxonomy annotate to phyloseq object<a class="anchor" aria-label="anchor" href="#converting-from-taxonomy-annotate-to-phyloseq-object"></a>
</h4>
<p><a href="https://joey711.github.io/phyloseq/" class="external-link">Phyloseq</a> is an R
package for exploring microbiome data. In particular, phyloseq provides
a number of visualizations (ordination, alpha diversity, heatmaps,
networks, etc.). The sourmashconsumr package provides a function,
<code><a href="../reference/from_taxonomy_annotate_to_phyloseq.html">from_taxonomy_annotate_to_phyloseq()</a></code>, to convert the
sourmash of sourmash taxonomy annotate into a phyloseq object.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gut_phyloseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/from_taxonomy_annotate_to_phyloseq.html">from_taxonomy_annotate_to_phyloseq</a></span><span class="op">(</span>taxonomy_annotate_df <span class="op">=</span> <span class="va">gut_taxonomy_annotate_df</span>, </span>
<span>                                                  metadata_df <span class="op">=</span> <span class="va">metadata</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>                                                    <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="st">"run_accessions"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gut_phyloseq</span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 437 taxa and 6 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 6 samples by 1 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 437 taxa by 8 taxonomic ranks ]</span></span></code></pre></div>
<p>This object can then be used with many of the built-in phyloseq
functions:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/plot_bar.html" class="external-link">plot_bar</a></span><span class="op">(</span><span class="va">gut_phyloseq</span><span class="op">)</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<p>In this case, the phyloseq object is built from the number of
abundance-weighted unique hashes (k-mers) that were identified in each
sample for each taxonomic lineage. This means that the scaled value will
impact the counts in the final phyloseq object. Therefore, it’s best to
build phyloseq objects from samples that were sketched and gathered
using the same scaled values.</p>
</div>
<div class="section level4">
<h4 id="converting-from-taxonomy-annotate-to-a-metacoder-object">Converting from taxonomy annotate to a metacoder object<a class="anchor" aria-label="anchor" href="#converting-from-taxonomy-annotate-to-a-metacoder-object"></a>
</h4>
<p><a href="https://grunwaldlab.github.io/metacoder_documentation/" class="external-link">Metacoder</a>
is an R package for visualizing heat trees from taxonomic data. The
sourmashconsumr package provides a function,
<code>from_taxonomy_to_metacoder()</code>, to convert the output of
sourmash taxonomy annotate to a metacoder object.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gut_metacoder</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/from_taxonomy_annotate_to_metacoder.html">from_taxonomy_annotate_to_metacoder</a></span><span class="op">(</span>taxonomy_annotate_df <span class="op">=</span> <span class="va">gut_taxonomy_annotate_df</span>, </span>
<span>                                                     intersect_bp_threshold <span class="op">=</span> <span class="fl">50000</span>, </span>
<span>                                                     tax_glom_level <span class="op">=</span> <span class="st">"order"</span>,</span>
<span>                                                     groups <span class="op">=</span> <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt; Summing per-taxon counts from 6 columns for 43 taxa</span></span>
<span><span class="co">#&gt; Calculating number of samples with a value greater than 0 for 6 columns in 2 groups for 43 observations</span></span></code></pre></div>
<p>This object can then be used with the built-in metacoder
visualizations like <code>heattree()</code>:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># generate a heat_tree plot with taxa from all samples, agglomerated to the order level</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="co"># This makes the plot appear the same each time it is run </span></span>
<span><span class="fu">metacoder</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/metacoder/man/heat_tree.html" class="external-link">heat_tree</a></span><span class="op">(</span><span class="va">gut_metacoder</span>, </span>
<span>                     node_label <span class="op">=</span> <span class="va">taxon_names</span>,</span>
<span>                     node_size <span class="op">=</span> <span class="va">n_obs</span>,</span>
<span>                     node_size_axis_label <span class="op">=</span> <span class="st">"k-mer abundance"</span>,</span>
<span>                     layout <span class="op">=</span> <span class="st">"davidson-harel"</span>, <span class="co"># The primary layout algorithm</span></span>
<span>                     initial_layout <span class="op">=</span> <span class="st">"reingold-tilford"</span><span class="op">)</span> <span class="co"># Node location algorithm </span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-37-1.png" width="700"></p>
<p>If your samples of interest come from two different groups, you can
also use the metacoder functions <code>compare_groups</code> and
<code>heat_tree_matrix</code> to visualize the differences between
groups. Note this only works with metacoder &gt;= 0.3.5.3. You may need
to install the development version of metacoder to access this
functionality. You can check which version of metacoder you have
installed using <code>packageVersion("metacoder")</code>. (At the time
this vignette was written, this fucntionality was not available on CRAN
so the below code was not executed.)</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use metacoder function compare_groups() to compare abundances between cd and nonibd gut microbiomes</span></span>
<span><span class="va">gut_metacoder</span><span class="op">$</span><span class="va">data</span><span class="op">$</span><span class="va">diff_table</span> <span class="op">&lt;-</span> <span class="fu">metacoder</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/metacoder/man/compare_groups.html" class="external-link">compare_groups</a></span><span class="op">(</span><span class="va">gut_metacoder</span>, data <span class="op">=</span> <span class="st">"tax_abund"</span>,</span>
<span>                                                           cols <span class="op">=</span> <span class="va">metadata</span><span class="op">$</span><span class="va">run_accessions</span>,</span>
<span>                                                           groups <span class="op">=</span> <span class="va">metadata</span><span class="op">$</span><span class="va">groups</span><span class="op">)</span></span>
<span><span class="co"># visualize the differences between groups using the metacoder function heat_tree_matrix()</span></span>
<span><span class="fu">metacoder</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/metacoder/man/heat_tree_matrix.html" class="external-link">heat_tree_matrix</a></span><span class="op">(</span><span class="va">gut_metacoder</span>,</span>
<span>                            data <span class="op">=</span> <span class="st">"diff_table"</span>,</span>
<span>                            node_size <span class="op">=</span> <span class="va">n_obs</span>,</span>
<span>                            node_label <span class="op">=</span> <span class="va">taxon_names</span>,</span>
<span>                            node_color <span class="op">=</span> <span class="va">log2_median_ratio</span>,</span>
<span>                            node_color_range <span class="op">=</span> <span class="fu">metacoder</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/metacoder/man/diverging_palette.html" class="external-link">diverging_palette</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                            node_color_trans <span class="op">=</span> <span class="st">"linear"</span>,</span>
<span>                            node_color_interval <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                            edge_color_interval <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                            node_size_axis_label <span class="op">=</span> <span class="st">"Abund of taxonomic lineage"</span>,</span>
<span>                            node_color_axis_label <span class="op">=</span> <span class="st">"Log2 ratio median proportions"</span><span class="op">)</span></span></code></pre></div>
<p>Like the phyloseq object, the metacoder object is built from the
number of abundance-weighted unique hashes (k-mers) that were identified
in each sample for each taxonomic lineage. This means that the scaled
value will impact the counts in the final metacoder object. Therefore,
it’s best to build metacoder objects from samples that were sketched and
gathered using the same scaled values.</p>
</div>
<div class="section level4">
<h4 id="detecting-whether-multiple-strains-of-the-same-species-are-present-in-a-single-metagenome">Detecting whether multiple strains of the same species are present
in a single metagenome<a class="anchor" aria-label="anchor" href="#detecting-whether-multiple-strains-of-the-same-species-are-present-in-a-single-metagenome"></a>
</h4>
<p>The sourmashconsumr package introduces a function to detect whether
multiple strains of the same species are likely present in a single
metagenome. It relies on <code>f_match</code> metric output by sourmash
gather as well as the taxonomic lineage (species) assigned by sourmash
taxonomy annotate. See the figure below for an explanation of how these
two pieces of information are combined to estimate whether multiple
strains are present. In this case, the reference database used in
sourmash gather is composed of four genomes of the same species (blue,
orange, green, and yellow). Each genome is represented as a sketch of
k-mers. Some k-mers are shared by all genomes in the database while
others are only present in some genomes. sourmash gather first finds the
best match between the k-mers in a metagenome and k-mers in a database.
In the example where only one genome is present (in purple), 0.8 (80%)
of the k-mers overlap with the blue genome so it is returned as the best
match. These k-mers are then subtracted and the search is repeated. The
orange genome is the next best match with 0.2 (20%) of the k-mers
overlapping, so it is returned next. The order the genome matches are
reported in is demarcated by a number next to the box that surrounds the
k-mers that matched a given genome. The fraction of the reference genome
found in the metagenome is one of the metrics reported in the sourmash
gather output (<code>f_match</code>). If only one strain (genome) of the
species is present in a metagenome, the summed fraction of all of the
genomes matched should not exceed ~1. When multiple strains are present
the first match found by sourmash gather may account for k-mers in both
strains. After summing the fraction of all matches, if the metagenome
was sequenced to sufficient depth, the summed fraction should exceed
~1.</p>
<p><img src="https://i.imgur.com/XPaYxqK.png" width="500"></p>
<p>Below we demonstrate how to use this approach on the results of
sourmash taxonomy annotate. We first read in a CSV file produced by
sourmash taxonomy annotate and filter it to contain valid taxonomic
lineages.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cami_taxonomy_annotate_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_taxonomy_annotate.html">read_taxonomy_annotate</a></span><span class="op">(</span><span class="st">"https://github.com/Arcadia-Science/2022-sourmashconsumr-validation/files/10513494/CAMI_low_vs_source_genomes.with-lineages.csv"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">species</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"unidentified"</span>, <span class="st">"unidentified virus"</span>, <span class="st">"unidentified plasmid"</span>, <span class="st">"Janthinobacterium species"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; File does not exist locally. Checking as url. If you did not pass a url, please provide a valid file path and retry. The Sys.glob() function might help if you're specifying multiple file paths.</span></span></code></pre></div>
<p>Then, we use the function
<code><a href="../reference/from_taxonomy_annotate_to_multi_strains.html">from_taxonomy_annotate_to_multi_strains()</a></code> to determine
which species likely have multiple strains present and to visualize
these species.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multi_strain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/from_taxonomy_annotate_to_multi_strains.html">from_taxonomy_annotate_to_multi_strains</a></span><span class="op">(</span><span class="va">cami_taxonomy_annotate_df</span><span class="op">)</span></span>
<span><span class="va">multi_strain</span><span class="op">$</span><span class="va">plt</span></span></code></pre></div>
<p><img src="sourmashconsumr_files/figure-html/unnamed-chunk-40-1.png" width="768"></p>
</div>
</div>
<div class="section level3">
<h3 id="upset-utility-functions">Upset utility functions<a class="anchor" aria-label="anchor" href="#upset-utility-functions"></a>
</h3>
<p>The sourmashconsumr package includes utility functions to create and
work with upset-formatted data frames. These data frames record the
presence-absence of the value that is intersected between many samples.
The columns are the sample names, and the rows are the object that is
shared by the samples. If a sample contained the object, a 1 appears in
the column; if it did not, a 0 appears.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gut_gather_upset_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 SRR5935765 SRR5936131 SRR5936197 SRR5946920 SRR5946923</span></span>
<span><span class="co">#&gt; GCF_002234575.2          1          1          0          1          0</span></span>
<span><span class="co">#&gt; GCF_001314995.1          1          1          1          1          1</span></span>
<span><span class="co">#&gt; GCF_000154205.1          1          1          1          1          1</span></span>
<span><span class="co">#&gt; GCF_000012825.1          1          1          1          1          1</span></span>
<span><span class="co">#&gt; GCF_002222615.2          1          1          1          1          1</span></span>
<span><span class="co">#&gt; GCF_000466485.1          1          1          0          1          0</span></span>
<span><span class="co">#&gt;                 SRR5947006</span></span>
<span><span class="co">#&gt; GCF_002234575.2          1</span></span>
<span><span class="co">#&gt; GCF_001314995.1          1</span></span>
<span><span class="co">#&gt; GCF_000154205.1          1</span></span>
<span><span class="co">#&gt; GCF_000012825.1          1</span></span>
<span><span class="co">#&gt; GCF_002222615.2          0</span></span>
<span><span class="co">#&gt; GCF_000466485.1          1</span></span></code></pre></div>
<p>The functions in the sourmashconsumr package pre-process or
post-process this data structure to extract information from it. The
<code><a href="../reference/from_list_to_upset_df.html">from_list_to_upset_df()</a></code> function is inspired by
<code>UpSetR::fromList()</code>. It creates the presence-absence data
frame seen above, but unlike the <code>UpSetR::fromList()</code>
function, it preserves the identity of the objects in the row names. The
rownames recording row identity can then act as an index to add metadata
to the presence-absence data frame that can then be visualized. In the
sourmashconsumr package, this was necessary to be able to color the the
upset plot bar graphs by group.</p>
<p>The functions <code><a href="../reference/from_upset_df_to_intersection_members.html">from_upset_df_to_intersection_members()</a></code>,
<code><a href="../reference/from_upset_df_to_intersection_summary.html">from_upset_df_to_intersection_summary()</a></code>, and
<code><a href="../reference/from_upset_df_to_intersections.html">from_upset_df_to_intersections()</a></code> take an upset-formatted
data frame as input and report on the identity of intersections between
groups of samples.</p>
<p><code><a href="../reference/from_upset_df_to_intersections.html">from_upset_df_to_intersections()</a></code> reports which samples
contain an object. The object is recorded as the row name, and the
intersection is reported as the sample names separated by an underscore.
If a sample doesn’t contain the object, that sample name is recorded as
a 0.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/from_upset_df_to_intersections.html">from_upset_df_to_intersections</a></span><span class="op">(</span><span class="va">gut_gather_upset_df</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                                      intersection</span></span>
<span><span class="co">#&gt; GCF_002234575.2                   SRR5935765_SRR5936131_0_SRR5946920_0_SRR5947006</span></span>
<span><span class="co">#&gt; GCF_001314995.1 SRR5935765_SRR5936131_SRR5936197_SRR5946920_SRR5946923_SRR5947006</span></span>
<span><span class="co">#&gt; GCF_000154205.1 SRR5935765_SRR5936131_SRR5936197_SRR5946920_SRR5946923_SRR5947006</span></span>
<span><span class="co">#&gt; GCF_000012825.1 SRR5935765_SRR5936131_SRR5936197_SRR5946920_SRR5946923_SRR5947006</span></span>
<span><span class="co">#&gt; GCF_002222615.2          SRR5935765_SRR5936131_SRR5936197_SRR5946920_SRR5946923_0</span></span>
<span><span class="co">#&gt; GCF_000466485.1                   SRR5935765_SRR5936131_0_SRR5946920_0_SRR5947006</span></span></code></pre></div>
<p>The order of the column names is consistent between objects.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/from_upset_df_to_intersections.html">from_upset_df_to_intersections</a></span><span class="op">(</span><span class="va">gut_gather_upset_df</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">intersection</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;                         intersection</span></span>
<span><span class="co">#&gt; GCA_000437275.1 0_0_0_0_0_SRR5947006</span></span>
<span><span class="co">#&gt; GCF_001404675.1 0_0_0_0_0_SRR5947006</span></span>
<span><span class="co">#&gt; GCF_002265625.1 0_0_0_0_0_SRR5947006</span></span>
<span><span class="co">#&gt; GCA_900555925.1 0_0_0_0_0_SRR5947006</span></span>
<span><span class="co">#&gt; GCA_900540595.1 0_0_0_0_0_SRR5947006</span></span>
<span><span class="co">#&gt; GCA_900542935.1 0_0_0_0_0_SRR5947006</span></span></code></pre></div>
<p><code><a href="../reference/from_upset_df_to_intersection_summary.html">from_upset_df_to_intersection_summary()</a></code> summarizes the
size of these intersections. Below, sample <code>SRR5946923</code> has
47 objects that are only observed in that sample. These intersection
sizes match those reported in the upset plot upper bar graph.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/from_upset_df_to_intersection_summary.html">from_upset_df_to_intersection_summary</a></span><span class="op">(</span><span class="va">gut_gather_upset_df</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">#&gt;   intersection                                                          n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                                             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 0_0_0_0_SRR5946923_0                                                 47</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 0_SRR5936131_0_0_0_0                                                 38</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 0_0_0_0_0_SRR5947006                                                 36</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 0_0_SRR5936197_0_SRR5946923_0                                        30</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> SRR5935765_SRR5936131_SRR5936197_SRR5946920_SRR5946923_SRR5947006    25</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 0_0_0_SRR5946920_0_0                                                 23</span></span></code></pre></div>
<p><code><a href="../reference/from_upset_df_to_intersection_members.html">from_upset_df_to_intersection_members()</a></code> reports the
identity of objects that are shared by each group.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">intersection_members</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/from_upset_df_to_intersection_members.html">from_upset_df_to_intersection_members</a></span><span class="op">(</span><span class="va">gut_gather_upset_df</span><span class="op">)</span></span></code></pre></div>
<p>The object that is output is a list, so it can be indexed into by
name. For example, if we were curious about the objects that were
observed in all samples, we could use the following:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">intersection_members</span><span class="op">$</span><span class="va">SRR5935765_SRR5936131_SRR5936197_SRR5946920_SRR5946923_SRR5947006</span></span>
<span><span class="co">#&gt;  [1] "GCF_001314995.1" "GCF_000154205.1" "GCF_000012825.1" "GCF_000239295.1"</span></span>
<span><span class="co">#&gt;  [5] "GCF_000484655.1" "GCA_900557355.1" "GCF_000210075.1" "GCA_900755095.1"</span></span>
<span><span class="co">#&gt;  [9] "GCF_014334015.1" "GCF_016908695.1" "GCF_005121165.2" "GCA_900760795.1"</span></span>
<span><span class="co">#&gt; [13] "GCF_009696065.1" "GCF_900106755.1" "GCF_013009555.1" "GCA_900761785.1"</span></span>
<span><span class="co">#&gt; [17] "GCA_902362375.1" "GCA_001915545.1" "GCF_009193295.2" "GCA_900758315.1"</span></span>
<span><span class="co">#&gt; [21] "GCF_003324185.1" "GCA_003573975.1" "GCA_003112755.1" "GCA_003584705.1"</span></span>
<span><span class="co">#&gt; [25] "GCA_900762515.1"</span></span></code></pre></div>
<p>The upset functions work on any upset-formatted data frame, not just
those that are created by the sourmashconsumr package. They will be most
meaningful if the upset data frame has row names.</p>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Taylor Reiter.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
